---
title: "Pea seed transcriptome processing"
author: "Yury V. Malovichko"
date: "November 20, 2019"
output: pdf_document
---

```{r setup, include=TRUE, warning=FALSE, message=FALSE}
library(dplyr)
library(magrittr)
library(tibble)
library(ggplot2)
library(topGO)
library(sleuth)
library(PoiClaClu)
library(RColorBrewer)
library(pheatmap)
library(ggbiplot)
library(lsa)
library(oposSOM)
library(UpSetR)

options(scipen = 5)
options(unzip = "internal", gtar = 'internal')
options(reutils.email = '271296251017a@gmail.com', reutils.show.headlines = 10, reutils.verbose.queries = T, reutils.test.remote = F)
options(OutDec= ".")
setwd(file.path('~', 'Pea_transcriptomics'))

```

## Transcriptome Annotation Uploading and Processing

First, we import three annotation sheets obtained independently from Trinotate, eggNOG and Mercator
```{r annotation_import echo=TRUE, warning=FALSE, message=FALSE}

annot_trinotate <-  read.table('novel_annotation/trinotate_annotation_report.csv', header = TRUE, sep = '\t', fill = TRUE, quote = "", stringsAsFactors = F) %>% 
  mutate_all(as.character())
annot_eggNOG <- read.table('novel_annotation/eggNOG_annotation.txt.emapper.annotations', header = TRUE, sep = '\t', fill = TRUE, quote = "", stringsAsFactors = F) %>% 
  mutate_all(as.character())
annot_mercator <- read.table('novel_annotation/mercator.results.txt', header = TRUE, sep = '\t', fill = TRUE, quote = "", stringsAsFactors = F) %>% 
  mutate_all(as.character()) %>% filter(IDENTIFIER %in% annot_trinotate$prot_id)

```

Now we create a simple function to convert full Trinotate GO annotation outputs into comma-separated strings:

```{r annotation_GO_processing, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
GO_deparser <- function(GO_entry) {
  GO_entry %>% strsplit('`') %>% lapply(function(x) x %>% strsplit('\\^') %>% lapply('[[', 1) %>% unlist() %>% paste(collapse = ','))
}

annot_trinotate$GO_BLAST <-  annot_trinotate %>% dplyr::select(gene_ontology_blast) %>% sapply(function(x) GO_deparser(x))
annot_trinotate$GO_BLAST[annot_trinotate$GO_BLAST == '.'] <- NA
annot_trinotate$PFAM_BLAST <-  annot_trinotate %>% dplyr::select(gene_ontology_pfam) %>% sapply(function(x) GO_deparser(x))
annot_trinotate$PFAM_BLAST[annot_trinotate$PFAM_BLAST == '.'] <- NA

for (i in 1:nrow(annot_trinotate)) {
  if (!is.na(annot_trinotate$PFAM_BLAST[i])){
  for (j in (annot_trinotate$PFAM_BLAST[i] %>% unlist() %>% strsplit(',') %>% unlist())) {
          if (!grepl(j, annot_trinotate$GO_BLAST[i])) annot_trinotate$GO_BLAST[i] <- paste(annot_trinotate$GO_BLAST[i], j, sep=',')
  }
          }
}

```

To further reinforce the completeness of the GO annotation we will also convert MapMan terms into GO terms using a Mercator MapMan-to-GO mapping:
```{r annotation_mercator_to_GO, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE}
MMtoGO <- read.table('mercator_mapping.tsv', sep = '\t', header = TRUE, stringsAsFactors = FALSE)
annot_mercator %<>% filter(IDENTIFIER != '')
annot_mercator$GO_BLAST_Mercator <- NA

for (i in 1:nrow(annot_mercator)) {
  annot_mercator$GO_BLAST_Mercator[i] <- MMtoGO[MMtoGO$BINNAME == annot_mercator$NAME[i], 'GO_TERM'] %>% paste(collapse = ',')
}

annot_mercator_new <- data.frame('protein' = annot_trinotate$prot_id, 'MapMan_term' = rep(NA, nrow(annot_trinotate)), 'GO_BLAST_Mercator' = rep(NA, nrow(annot_trinotate)))

for (i in 1:nrow(annot_mercator_new)) {
  annot_mercator_new$MapMan_term[i] <- annot_mercator[annot_mercator$IDENTIFIER == annot_mercator_new$protein[i], 'NAME'] %>% paste(collapse = ';')
annot_mercator_new$GO_BLAST_Mercator[i] <- annot_mercator[annot_mercator$IDENTIFIER == annot_mercator_new$protein[i], 'GO_BLAST_Mercator']
  }

annot_mercator_new[annot_mercator_new$GO_BLAST_Mercator == '',]$GO_BLAST_Mercator <- NA

```

To sum Trinotate GO annotations with those produced by eggNOG and MapMan-to-GO mapping we will check for each of the external term whether it is present in the Trinotatr output:

```{r GO_gathering echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}
for (i in 1:nrow(annot_trinotate)) {
  for (j in annot_mercator_new[annot_mercator_new$protein == annot_trinotate$prot_id[i],]$GO_BLAST_Mercator) { 
    if (is.na(annot_trinotate$GO_BLAST[i])) annot_trinotate$GO_BLAST[i] <- j
    else if (!is.na(j)) {
      for (k in strsplit(j,',') %>% unlist()) {
     if (!grepl(k, annot_trinotate$GO_BLAST[i])) {
annot_trinotate$GO_BLAST[i] <- paste(annot_trinotate$GO_BLAST[i], k, sep = ',')
    }    
}    
}  
}
}


for (i in 1:nrow(annot_trinotate)) {
  for (j in annot_eggNOG[annot_eggNOG$query_name == annot_trinotate$prot_id[i],]$GO_BLAST) {
    if (!is.na(j)) {
    for (k in strsplit(j,',') %>% unlist()) {

    if (!is.na(k) & !grepl(k, annot_trinotate$GO_BLAST[i])) {
      annot_trinotate$GO_BLAST[i] <- paste(annot_trinotate$GO_BLAST[i], k, sep = ',')
    }
      }
      }
  }
}

```

The final annotation sheet looks like the following:
```{r annotation_merging, echo=TRUE, eval=TRUE, message=FALSE, error=FALSE}
annot_full_ext <- cbind(annot_trinotate, annot_mercator_new[match(annot_trinotate$prot_id, annot_mercator_new$protein), c('MapMan_term', 'GO_BLAST_Mercator')],
                    annot_eggNOG[match(annot_trinotate$prot_id, annot_eggNOG$query_name), c('KEGG_ko', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction')])
annot_full_ext[annot_full_ext == '.'] <- NA
annot_full_ext[annot_full_ext == ''] <- NA

data.table::fwrite(list(sapply(annot_full_ext$plants_seqs_BLASTX, function(x)
  x %>% strsplit('\\^') %>% unlist() %>% extract2(1))), file = file.path('BLAST_decoding','BLASTX_identifiers'))

data.table::fwrite(list(sapply(annot_full_ext$plants_seqs_BLASTP, function(x)
  x %>% strsplit('\\^') %>% unlist() %>% extract2(1))), file = file.path('BLAST_decoding','BLASTP_identifiers'))

annot_full_ext %<>% mutate(BLASTP_names = read.table(file.path('BLAST_decoding', 'BLASTP_descriptions.txt'), stringsAsFactors = F, sep = '\t', comment.char = '', quote = '')$V1,
      BLASTX_names = read.table(file.path('BLAST_decoding', 'BLASTX_descriptions.txt'), stringsAsFactors = F, sep = '\t', comment.char = '', quote = '')$V1)


write.table(apply(annot_full_ext, 2, as.character), file.path('novel_publication_materials', 'TableS2.tsv'), sep = '\t', col.names = T, row.names = F)

annot_full_ext %>% filter(is.na(plants_seqs_BLASTP), is.na(plants_seqs_BLASTX)) %>% dplyr::select(transcript_id) %>% write.table(file.path('unblasted_values.txt'), col.names = F, row.names = F, sep = '\t')

annot_full_ext %>% filter(is.na(plants_seqs_BLASTP), is.na(plants_seqs_BLASTX)) %>% dplyr::select(prot_id) %>% write.table(file.path('unblasted_proteins.txt'), col.names = F, row.names = F, sep = '\t')


```

## Combined assembly annotation
The same approach is applied to the combined assembly comprising of those related to Sprint-2 and published by <> et al. (2017), For the sake of time economy only contigs borrowed from the external assemblies have been annotated separately, and the information on those deduced from Sprint-2 transcriptome is retrieved from the respective annotation spreadshit.

```{r combined_annotation, echo=TRUE, eval=TRUE, warning=FALSE, error=FALSE, message=FALSE}

trinity_names <-scan(file.path('combined_annotations', 'Trinity_contigs_for_combined_assembly.txt'), character(), quote='') 

external_annotation <- read.table(file.path('combined_annotations', 'external_contigs_only.xls'), sep = '\t', header = T, quote = '', stringsAsFactors = F)

data.table::fwrite(list(sapply(external_annotation$plant_seqs_BLASTX, function(x)
  x %>% strsplit('\\^') %>% unlist() %>% extract2(1))), file = file.path('combined_annotations','BLASTX_identifiers.txt'))

data.table::fwrite(list(sapply(external_annotation$plant_seqs_BLASTP, function(x)
  x %>% strsplit('\\^') %>% unlist() %>% extract2(1))), file = file.path('combined_annotations','BLASTP_identifiers.txt'))

external_annotation$BLASTP_names <- scan(file.path('combined_annotations', 'BLASTP_names.txt'), character(), quote = '', sep = '\n')
external_annotation$BLASTX_names <- scan(file.path('combined_annotations', 'BLASTX_names.txt'), character(), quote = '', sep = '\n')

external_annotation$GO_BLAST <-  external_annotation %>% dplyr::select(gene_ontology_blast) %>% sapply(function(x) GO_deparser(x)) %>% unlist()
external_annotation$GO_BLAST[external_annotation$GO_BLAST == '.'] <- NA
external_annotation$PFAM_BLAST <-  external_annotation %>% dplyr::select(gene_ontology_pfam) %>% sapply(function(x) GO_deparser(x)) %>% unlist()
external_annotation$PFAM_BLAST[external_annotation$PFAM_BLAST == '.'] <- NA

for (i in 1:nrow(external_annotation)) {
  if (!is.na(external_annotation$PFAM_BLAST[i])){
  for (j in (external_annotation$PFAM_BLAST[i] %>% unlist() %>% strsplit(',') %>% unlist())) {
          if (!grepl(j, external_annotation$GO_BLAST[i])) external_annotation$GO_BLAST[i] <- paste(external_annotation$GO_BLAST[i], j, sep=',')
  }
          }
}



annot_mercator_ext <- read.table('combined_annotations/mercator.results.txt', header = TRUE, sep = '\t', fill = TRUE, quote = "", stringsAsFactors = F) %>%  mutate_all(as.character()) %>% filter(IDENTIFIER %in% external_annotation$prot_id)

MMtoGO <- read.table('mercator_mapping.tsv', sep = '\t', header = TRUE, stringsAsFactors = FALSE)
annot_mercator_ext %<>% filter(IDENTIFIER != '')
annot_mercator_ext$GO_BLAST_Mercator <- NA

for (i in 1:nrow(annot_mercator_ext)) {
  annot_mercator_ext$GO_BLAST_Mercator[i] <- MMtoGO[MMtoGO$BINNAME == annot_mercator_ext$NAME[i], 'GO_TERM'] %>% paste(collapse = ',')
}

annot_mercator_ext_new <- data.frame('protein' = external_annotation$prot_id, 'MapMan_term' = rep(NA, nrow(external_annotation)), 'GO_BLAST_Mercator' = rep(NA, nrow(external_annotation)))

for (i in 1:nrow(annot_mercator_ext_new)) {
  annot_mercator_ext_new$MapMan_term[i] <- annot_mercator_ext[annot_mercator_ext$IDENTIFIER == annot_mercator_ext_new$protein[i], 'NAME'] %>% paste(collapse = ';')
annot_mercator_ext_new$GO_BLAST_Mercator[i] <- annot_mercator_ext[annot_mercator_ext$IDENTIFIER == annot_mercator_ext_new$protein[i], 'GO_BLAST_Mercator'] %>% paste(collapse = ',')
  }

annot_mercator_ext_new[annot_mercator_ext_new$GO_BLAST_Mercator == '',]$GO_BLAST_Mercator <- NA

annot_eggNOG_ext <- read.table('combined_annotations/eggNOG_headless.tsv', header = TRUE, sep = '\t', fill = TRUE, quote = "", stringsAsFactors = F) %>% 
  mutate_all(as.character())
annot_eggNOG_ext[annot_eggNOG_ext$GO_BLAST == '',] <- NA
annot_eggNOG_ext %>% filter(!is.na(GO_BLAST)) %>% nrow()



for (i in 1:nrow(external_annotation)) {
  for (j in annot_mercator_ext_new[annot_mercator_ext_new$protein == external_annotation$prot_id[i],]$GO_BLAST_Mercator) { 
    if (is.na(external_annotation$GO_BLAST[i])) external_annotation$GO_BLAST[i] <- j
    else if (!is.na(j)) {
      for (k in strsplit(j,',') %>% unlist()) {
     if (!grepl(k, external_annotation$GO_BLAST[i])) {
external_annotation$GO_BLAST[i] <- paste(external_annotation$GO_BLAST[i], k, sep = ',')
    }    
}    
}  
}
}


for (i in 1:nrow(external_annotation)) {
  for (j in annot_eggNOG_ext[annot_eggNOG_ext$query_name == external_annotation$prot_id[i],]$GO_BLAST) {
    if (!is.na(j)) {
      if (is.na(external_annotation$GO_BLAST[i])) external_annotation$GO_BLAST[i] <- j
    else for (k in strsplit(j,',') %>% unlist()) {

    if (!is.na(k) & !grepl(k, external_annotation$GO_BLAST[i])) {
      external_annotation$GO_BLAST[i] <- paste(external_annotation$GO_BLAST[i], k, sep = ',')
    }
      }
      }
  }
}


annot_full_ext <-  cbind(external_annotation, annot_mercator_ext_new[match(external_annotation$prot_id, annot_mercator_ext_new$protein), c('MapMan_term', 'GO_BLAST_Mercator')], annot_eggNOG_ext[match(external_annotation$prot_id, annot_eggNOG_ext$query_name), c('KEGG_ko', 'KEGG_Pathway', 'KEGG_Module', 'KEGG_Reaction')])
annot_full_ext <- annot_full_ext[,c(1:15, 18:25, 16, 17)]
colnames(annot_full_ext) <- colnames(annot_full_ext)

annot_full_ext <-  rbind(annot_full_ext %>% filter(transcript_id %in% trinity_names), annot_full_ext)
annot_full_ext[annot_full_ext == ','] <- NA
annot_full_ext$GO_BLAST %<>% unlist()
annot_full_ext$PFAM_BLAST %<>% unlist()
write.table(annot_full_ext, file.path('very_last_supplementary_materials', 'supplementary_tables', 'Table_S4.tsv'),
            col.names = T, row.names = F, sep = '\t')
```

## GO Over-Representation Functions
GO over-representation tests are the workhorses of downstream DE analysis, so the prepared functions would come in handy:

```{r GO_functions, echo=TRUE, eval=TRUE, message=FALSE, error=FALSE}

full_mapping <- annot_full_ext %>% dplyr::select(transcript_id, GO_BLAST) %>% mutate(transcript_id = as.character(transcript_id),
                                                                            GO_BLAST = as.character(GO_BLAST))
full_mapping <- full_mapping %>% filter(!is.na(GO_BLAST))
full_mapping$GO_BLAST <- sapply(full_mapping$GO_BLAST, function(x) strsplit(x, ','))
names(full_mapping$GO_BLAST) <- full_mapping$transcript_id
full_mapping <- full_mapping$GO_BLAST
geneUniverse <- full_mapping %>% names()

newFuncadelicGo <- function(x, database, mode, intGenes, fm, cutoff = 0.01) {
  myGOdata <- new("topGOdata", 
                  description = '', ontology = database,
                  allGenes=x,
                  geneSel=selector,
                  annot=annFUN.gene2GO, gene2GO=fm, nodeSize=10)
  resultFisher <- runTest(myGOdata, algorithm = "weight01", statistic = "fisher")

  print(myGOdata)
  
  allRes <- GenTable(myGOdata, weight01Fisher = resultFisher,
                     orderBy = "weight01Fisher", ranksOf = "weight01Fisher", topNodes = 70, numChar = 200)
  # allRes[is.na(allRes)] <- 0
  if (mode=='table') {
    allRes$Percentage_of_Differentially_Expressed_Genes_from_number_of_DEGs <- ((allRes$Significant / nrow(intGenes))*100) %>% round(2)
    allRes$Percentage_of_Differentially_Expressed_Genes_from_Number_Of_Annotated_Genes <- ((allRes$Significant / allRes$Annotated)*100) %>% round(2)
    allRes <- allRes %>% filter(weight01Fisher < cutoff)
    return(allRes)
  } else if (mode=='pvals'){
    retlist <- (allRes %>% filter(weight01Fisher < cutoff))$weight01Fisher
    names(retlist) <- (allRes %>% filter(weight01Fisher < cutoff))$GO.ID
    return(retlist)
  }
  else if(mode=='graph') {
    GoGraph <- showSigOfNodes(myGOdata, score(resultFisher), firstSigNodes = allRes %>% filter(weight01Fisher < cutoff) %>% nrow(), useInfo = "all")$dag
    return(GoGraph)
  }
}

iterating_function_GO <- function(subset, ont, cutoff, f_m, gU, make_table = T){
  dummy <- subset %>% dplyr::select(target_id) %>% mutate(target_id = as.character(target_id)) %>% unlist()
  geneList <- factor(as.integer(gU %in% dummy))
  names(geneList) <- gU
  name_val <- paste(ont, substitute(subset) %>% as.character(), sep = '_')
  if (make_table) assign(name_val, newFuncadelicGo(geneList, ont, mode = 'table', intGenes=subset, fm = f_m, cutoff = cutoff), envir = .GlobalEnv)
}

aspects <- c('BP', 'CC', 'MF')


```


Down the road manual tracking of functional categories based on the GO annotation will be required. Though these lists seem way too bulk, they will come in handy for both GO over-representation and scatter plot mapping:

```{r GO_lists, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE, error=FALSE}

TF_GO <- c('GO:0003700', 'GO:0001217', 'GO:0001216', 'GO:0051091', 'GO:0043433', 'GO:0051090', 'GO:0098531',
           'GO:0034246', 'GO:0005667', 'GO:0000981', 'GO:0003712', 'GO:0003714', 'GO:0003713',
           'GO:0006325', 'GO:1990700', 'GO:0006338', 'GO:0035092', 'GO:0070827', 'GO:0016569',
           'GO:0070828', 'GO:0090202', 'GO:0006333', 'GO:1905269', 'GO:1902275', 'GO:1905268',
           'GO:0034401', 'GO:0001301', 'GO:0034728', 'GO:0039525', 'GO:0061641', 'GO:0006355',
           'GO:0045892', 'GO:0003677')

lipids <- c('GO:0005811', 'GO:0055089', 'GO:0015908', 'GO:0006631', 'GO:0006633', 'GO:0030497', 'GO:0019915')
storage_proteins <- c('GO:0030497', 'GO:0034422', 'GO:0032578', 'GO:0045735', 'GO:0000326', 'GO:1990019')
### GO:0043556 - regulation of translation in response to oxidative stress; GO:0032939 - positive, GO:0032938 - negative
### GO:1990497 - regulation to cytoplasmic translation in response to stress
### GO:0032057 - negative modulation of translational initiation in response to stress

hormones <- c('GO:0016707', 'GO:0010336', 'GO:001047', 'GO:0045487', 'GO:0009739', 'GO:0009686', 'GO:0009685',
                 'GO:0033470', 'GO:0033469', 'GO:0071370', 'GO:0010371', 'GO:0010372', 'GO:0010373', 'GO:0010331',
                 'GO:0045544', 'GO:0102713', 'GO:1905201', 'GO:0047927', 'GO:0102111', 'GO:0102119', 'GO:0102118',
                 'GO:0102117', 'GO:0102122', 'GO:0045543', 'GO:0102714', 'GO:0102716', 'GO:0102715', 'GO:0102712',
                 'GO:0102711', 'GO:0016707', 'GO:0102653', 'GO:0102652', 'GO:0102663', 'GO:0102924', 'GO:0051779', 'GO:0047928',
                 'GO:0010341', 'GO:0102123', 'GO:0102125', 'GO:0102124', 'GO:0102738', 'GO:0102739', 'GO:0103057', 'GO:0103056',
                 'GO:0052634', 'GO:0052635', 'GO:0103010', 'GO:0102972', 'GO:0103054', 'GO:0102755',
                 'GO:0080168', 'GO:1902265', 'GO:0010427', 'GO:0009687', 'GO:0010294', 'GO:0010293',
                 'GO:1902266', 'GO:0046345', 'GO:0009737', 'GO:0009688', 'GO:0090440', 'GO:0010295', 'GO:0009724', 'GO:0009738',
                 'GO:0071215', 'GO:0010115', 'GO:0090359', 'GO:0051993', 'GO:1902418', 'GO:0010116', 'GO:0009787', 'GO:1902417',
                 'GO:0009789', 'GO:0009788', 'GO:0075343', 'GO:1901527', 'GO:1990218',
                 'GO:0060918', 'GO:0060919', 'GO:0010011', 'GO:0010252', 'GO:0010315', 'GO:0009850', 'GO:0038198', 'GO:0080162',
                 'GO:0009733', 'GO:0009852', 'GO:0009851', 'GO:0010541', 'GO:0010540', 'GO:0009926', 'GO:0010249', 'GO:0010328',
                 'GO:0080161', 'GO:0009721', 'GO:0009734', 'GO:0009921', 'GO:0090354', 'GO:2000012', 'GO:0071365', 'GO:0010329',
                 'GO:0010600', 'GO:0090355', 'GO:0090356', 'GO:0010601', 'GO:0010928', 'GO:1901703', 'GO:0010930', 'GO:0010929',
                 'GO:0060774', 'GO:0090015', 'GO:0044032', 'GO:1990210', 'GO:0044846', 'GO:0009672',
                 'GO:0010184', 'GO:0044373', 'GO:0009690', 'GO:0009735', 'GO:0009691', 'GO:0009823', 'GO:0009884', 'GO:0019139',
                 'GO:0009722', 'GO:0009736', 'GO:0080062', 'GO:1903856', 'GO:0047807', 'GO:0071368', 'GO:0080036', 'GO:1903857',
                 'GO:0001647', 'GO:0009885', 'GO:0080037', 'GO:0080038', 'GO:1990223', 'GO:0009824', 'GO:0050447',
                 'GO:0051740', 'GO:0042457', 'GO:0038199', 'GO:0009723', 'GO:0009693', 'GO:0009692', 'GO:0009727', 'GO:0009873',
                 'GO:0038200', 'GO:0071369', 'GO:0010364', 'GO:0075022', 'GO:0010365', 'GO:0010366', 'GO:0010104', 'GO:0102276',
                 'GO:0010105', 'GO:0009861', 'GO:0009866', 'GO:0052021', 'GO:0052084', 'GO:1990212', 'GO:0052076',
                 'GO:0052005', 'GO:0043272', 'GO:0009871', 'GO:0009868', 'GO:0032260', 'GO:0052441', 'GO:0009815', 'GO:0004085',
                 'GO:0052276',
                 'GO:0030795', 'GO:0080123', 'GO:0102078', 'GO:0080032', 'GO:0009867', 'GO:0120091', 'GO:0009694', 'GO:0009753',
                 'GO:0009695', 'GO:0009754', 'GO:0009754', 'GO:0080140', 'GO:0080141', 'GO:0071395', 'GO:2000022', 'GO:0009861',
                 'GO:0009864', 'GO:0052022', 'GO:0052088', 'GO:1990211', 'GO:0052075', 'GO:0052068', 'GO:0043272', 'GO:0009871',
                 'GO:0009868', 'GO:0052073', 'GO:0032260', 'GO:0080032', 'GO:0052480', 'GO:0030795', 'GO:0080123',
                 'GO:0090411', 'GO:0010268', 'GO:0016131', 'GO:0016132', 'GO:0016133', 'GO:0080118', 'GO:0009741', 'GO:0009742',
                 'GO:0009729', 'GO:0071367', 'GO:0010422', 'GO:1900457', 'GO:2000488', 'GO:0010423', 'GO:1900458', 'GO:1900459',
                 'GO:1901149', 'GO:0009696', 'GO:0046244', 'GO:0009751', 'GO:0009697', 'GO:0009752', 'GO:0009863', 'GO:0010337',
                 'GO:0009627', 'GO:0080142', 'GO:0052639', 'GO:0052640', 'GO:0071446', 'GO:2000031', 'GO:0080151', 'GO:0009862',
                 'GO:0052023', 'GO:0010679', 'GO:0052089', 'GO:1990213', 'GO:0052074', 'GO:0052004', 'GO:0052081', 'GO:0052072',
                 'GO:0052003', 'GO:0052284', 'GO:0052253', 'GO:0052272', 'GO:0080031', 'GO:0018658', 'GO:0052624', 'GO:1901672',
                 'GO:1902347', 'GO:1902348', 'GO:1901601', 'GO:1901600')

translation <- c('GO:0006412', 'GO:0002181', 'GO:0002188', 'GO:0032543', 'GO:0032544', 'GO:0030371', 'GO:0006417',
                 'GO:0070992', 'GO:0070993', 'GO:0008494', 'GO:0045182', 'GO:1903502', 'GO:0070126', 'GO:0070124',
                 'GO:0070125', 'GO:0017148', 'GO:0018444', 'GO:0031369', 'GO:0070129', 'GO:0045727', 'GO:0008079',
                 'GO:2000765', 'GO:0003743', 'GO:0003746', 'GO:0003747', 'GO:0097167', 'GO:0061770', 'GO:0044207',
                 'GO:0006418', 'GO:0045974', 'GO:0001731', 'GO:0070130', 'GO:0070131', 'GO:0008135', 'GO:0035278',
                 'GO:2000767', 'GO:2000766', 'GO:0046011', 'GO:1905082', 'GO:0006414', 'GO:0070132', 'GO:0001677',
                 'GO:0040033', 'GO:0005851', 'GO:0005850', 'GO:0005853', 'GO:0005852', 'GO:0090079', 'GO:0016281',
                 'GO:0016150', 'GO:0016149', 'GO:0045975', 'GO:0001732', 'GO:0070127', 'GO:0043143', 'GO:0043614',
                 'GO:0070133', 'GO:0070134', 'GO:0000900', 'GO:0043555', 'GO:0039606', 'GO:1905173', 'GO:0097010',
                 'GO:1905143', 'GO:1905616', 'GO:0004694', 'GO:0045183', 'GO:0070196', 'GO:1901193', 'GO:0000901',
                 'GO:0071541', 'GO:0071540', 'GO:0006413', 'GO:1901190', 'GO:1990500', 'GO:0043556', 'GO:0043557',
                 'GO:0070549', 'GO:1905618', 'GO:1905617', 'GO:0070321', 'GO:1901194', 'GO:1901195', 'GO:1990497',
                 'GO:0032055', 'GO:0032056', 'GO:0071891', 'GO:0043558', 'GO:0080149', 'GO:0006415', 'GO:0036490',
                 'GO:0032939', 'GO:0032938', 'GO:1905535', 'GO:0070322', 'GO:0070323', 'GO:1901192', 'GO:1901191',
                 'GO:1904803', 'GO:0032061', 'GO:0032062', 'GO:0010892', 'GO:0036493', 'GO:0036491', 'GO:1905537',
                 'GO:1905536', 'GO:1902010', 'GO:0036494', 'GO:0036495', 'GO:0006451', 'GO:0006452', 'GO:0009386',
                 'GO:0036497', 'GO:0036492', 'GO:0002182', 'GO:0002183', 'GO:0002184', 'GO:0002190', 'GO:0002191',
                 'GO:1990145', 'GO:0075522', 'GO:0006446', 'GO:0006450', 'GO:0006448', 'GO:0006449', 'GO:2001124',
                 'GO:0032057', 'GO:0045948', 'GO:0045947', 'GO:0045904', 'GO:0045905', 'GO:0045902', 'GO:0045903',
                 'GO:0045900', 'GO:0045901', 'GO:1903674', 'GO:1903677', 'GO:1904688', 'GO:1990580', 'GO:1900247',
                 'GO:2001126', 'GO:2001125', 'GO:0140018', 'GO:0032063', 'GO:1903916', 'GO:0032058', 'GO:1905084',
                 'GO:1905083', 'GO:0002192', 'GO:0006447', 'GO:0106074', 'GO:0110019', 'GO:0110017', 'GO:0110018',
                 'GO:1903675', 'GO:1903678', 'GO:1903679', 'GO:1903676', 'GO:1904690', 'GO:1904689', 'GO:1900249',
                 'GO:1900248', 'GO:0097622', 'GO:0071263', 'GO:0032064', 'GO:0045993', 'GO:0045994', 'GO:0036499',
                 'GO:0071264', 'GO:1903270', 'GO:0036496', 'GO:0071262', 'GO:0010998', 'GO:0043561', 'GO:1903272',
                 'GO:1903271', 'GO:1990611', 'GO:1990625', 'GO:1903917', 'GO:1903912', 'GO:0007571', 'GO:0010998')

repair <- c('GO:0097551', 'GO:0036297', 'GO:000630', 'GO:0006281', 'GO:0006298', 'GO:0000719', 'GO:0000725',
                'GO:0006284', 'GO:0006289', 'GO:0006290', 'GO:0043504', 'GO:0001778', 'GO:1990391', 'GO:1990516',
                'GO:0010206', 'GO:0009380', 'GO:0000710', 'GO:0000726', 'GO:0032300', 'GO:1990710', 'GO:0010213',
                'GO:0006302', 'GO:0006282', 'GO:0090735', 'GO:0070914', 'GO:0000012', 'GO:0000109', 'GO:0000711',
                'GO:0032423', 'GO:0032404', 'GO:0051103', 'GO:0097552', 'GO:0006283', 'GO:0006288', 'GO:0006287',
                'GO:0033683', 'GO:0042275', 'GO:1905051', 'GO:0070911', 'GO:1903823', 'GO:1905684', 'GO:0045739',
                'GO:0045738', 'GO:2000819', 'GO:0097520', 'GO:0036298', 'GO:0032425', 'GO:0032424', 'GO:0006294',
                'GO:0003684', 'GO:1990414', 'GO:0000716', 'GO:0000715', 'GO:0006307', 'GO:0006285', 'GO:0006293',
                'GO:0006297', 'GO:1905053', 'GO:1905052', 'GO:0000111', 'GO:0000110', 'GO:0000113', 'GO:0000112',
                'GO:1905686', 'GO:1905685', 'GO:1903516', 'GO:1902113', 'GO:2000779', 'GO:1990598', 'GO:0097698',
                'GO:1990731', 'GO:0036299', 'GO:0000731', 'GO:0000718', 'GO:0000717', 'GO:0140274', 'GO:1990249',
                'GO:0098783', 'GO:0090262', 'GO:0098504', 'GO:1903517', 'GO:1903518', 'GO:1990396', 'GO:2000780',
                'GO:2000781', 'GO:0000724', 'GO:0000720', 'GO:0140273', 'GO:0006295', 'GO:0006296', 'GO:0006303',
                'GO:0006286', 'GO:1904161', 'GO:0090699', 'GO:1903824', 'GO:0045002', 'GO:1990918', 'GO:0000727',
                'GO:0010776', 'GO:0010777', 'GO:1990250', 'GO:1903110', 'GO:1901255', 'GO:0045003', 'GO:0097681',
                'GO:0097680', 'GO:0061306', 'GO:0000734', 'GO:0010779', 'GO:0010778', 'GO:0010569', 'GO:0043150',
                'GO:1904162', 'GO:1903111', 'GO:1903112', 'GO:1905168', 'GO:1901591', 'GO:2000042', 'GO:2001032',
                'GO:0097510', 'GO:0043151', 'GO:0042276', 'GO:1901291', 'GO:1901592', 'GO:2001033', 'GO:2001034',
                'GO:1902346', 'GO:0043765', 'GO:0043739', 'GO:0070716', 'GO:0100026', 'GO:0061674', 'GO:0000736',
                'GO:0010792')

starch <- c('GO:0043036', 'GO:2001070', 'GO:0005982', 'GO:0062052', 'GO:0005983', 'GO:0009011', 'GO:0009569',
            'GO:0009568', 'GO:0019252', 'GO:0044654', 'GO:0004373', 'GO:0102218', 'GO:0044570', 'GO:2000904',
            'GO:0102502', 'GO:0033840', 'GO:2000881', 'GO:0010581', 'GO:0044574', 'GO:2000905', 'GO:2000906',
            'GO:1904160', 'GO:2000465', 'GO:2000882', 'GO:2000883', 'GO:0102222', 'GO:2000467', 'GO:2000466',
            'GO:1900512', 'GO:1900513', 'GO:1900514', 'GO:0035956', 'GO:0050521', 'GO:0003844')

desiccation <- c('GO:0009269', 'GO:0071465', 'GO:0097439', 'GO:0048700', 'GO:0072515')

postmod <- c('GO:0043687', 'GO:1901873', 'GO:0036211', 'GO:0042160', 'GO:0031179', 'GO:0030047', 'GO:0018208', 'GO:0018207',
             'GO:0018209', 'GO:0018211', 'GO:0018210', 'GO:0018213', 'GO:0018212', 'GO:0018202', 'GO:0018201', 'GO:0018204',
             'GO:0018203', 'GO:0018206', 'GO:0018205', 'GO:0018194', 'GO:0018196', 'GO:0018195', 'GO:0018198', 'GO:0018199',
             'GO:0050844', 'GO:1901875', 'GO:1901874', 'GO:0072580', 'GO:0043686', 'GO:0006464', 'GO:0018200', 'GO:0018193',
             'GO:0018197', 'GO:0071587', 'GO:0140030', 'GO:0035610', 'GO:0036210', 'GO:0034421', 'GO:0006497', 'GO:0018410',
             'GO:0031400', 'GO:0031401', 'GO:0031365', 'GO:0032446', 'GO:0140035', 'GO:0070647', 'GO:0098823', 'GO:1903059',
             'GO:1903320', 'GO:1903322', 'GO:1903321', 'GO:1903061', 'GO:1903060', 'GO:0031289', 'GO:0006468', 'GO:0007258',
             'GO:0035404', 'GO:0035405', 'GO:0035406', 'GO:1990164', 'GO:0018109', 'GO:0018108', 'GO:0018105', 'GO:0018107',
             'GO:0018106', 'GO:0018218', 'GO:1990245', 'GO:0043989', 'GO:0043988', 'GO:0043987', 'GO:0043990', 'GO:0043991',
             'GO:0043538', 'GO:0035408', 'GO:0035409', 'GO:0035407', 'GO:0035978', 'GO:0033127', 'GO:0018217', 'GO:0072370',
             'GO:0072355', 'GO:0001932', 'GO:0022828', 'GO:0140031', 'GO:0042501', 'GO:0050730', 'GO:0033135', 'GO:0033129',
             'GO:0033128', 'GO:0001933', 'GO:0001934', 'GO:0023015', 'GO:0023014', 'GO:0023016', 'GO:1990853', 'GO:0010799',
             'GO:0002030', 'GO:0050731', 'GO:0050732', 'GO:0033137', 'GO:0033138', 'GO:2000281', 'GO:0010801', 'GO:0010800',
             'GO:2000775', 'GO:0036492', 'GO:0010998', 'GO:1904324', 'GO:1904325', 'GO:0100002', 'GO:0060734', 'GO:0060733',
             'GO:1903125', 'GO:1903912', 'GO:1901408', 'GO:1901409', 'GO:0071619', 'GO:0071620', 'GO:0044387', 'GO:2000751',
             'GO:2001163', 'GO:2000817', 'GO:2001164', 'GO:2001165', 'GO:1903654', 'GO:1903655', 'GO:1900018', 'GO:0019901',
             'GO:0004672', 'GO:1902911', 'GO:0032147', 'GO:0019887', 'GO:0006486', 'GO:0018280', 'GO:0033576', 'GO:0033578',
             'GO:0042076', 'GO:0018103', 'GO:0006487', 'GO:0006493', 'GO:0060049', 'GO:0140032', 'GO:0033577', 'GO:0033575',
             'GO:0060050', 'GO:0060051', 'GO:0042543', 'GO:0018317', 'GO:0018279', 'GO:0018258', 'GO:0018240', 'GO:0018242',
             'GO:0018241', 'GO:0018244', 'GO:0018243', 'GO:0018245', 'GO:0042077', 'GO:0090283', 'GO:1904098', 'GO:1904100',
             'GO:0090285', 'GO:0090284', 'GO:1904099', 'GO:0035629', 'GO:0018406', 'GO:0032147', 'GO:0043549', 'GO:1990782')

degradation <- c('GO:0016574', 'GO:0016567', 'GO:0007014', 'GO:0033522', 'GO:0033523', 'GO:0140372', 'GO:0070534', 'GO:0085020', 'GO:0070979',
                 'GO:0031396', 'GO:0070936', 'GO:0033182', 'GO:0035519', 'GO:1990390', 'GO:0036351', 'GO:0036352', 'GO:0140373', 'GO:0044314',
                 'GO:0070535', 'GO:0031398', 'GO:0031397', 'GO:0033184', 'GO:0033183', 'GO:2001166', 'GO:1900044', 'GO:0061945', 'GO:2001167',
                 'GO:2001168', 'GO:0140035', 'GO:0039648', 'GO:1901314', 'GO:1902523', 'GO:1902524', 'GO:1900045', 'GO:0061944', 'GO:0071894',
                 'GO:0075346', 'GO:1901316', 'GO:1901315', 'GO:1990756', 'GO:2001173', 'GO:2001174', 'GO:2001175', 'GO:0039542', 'GO:0051865',
                 'GO:1904666', 'GO:1902498', 'GO:0061630', 'GO:2000060', 'GO:2000059', 'GO:0006511', 'GO:2000058', 'GO:1904668', 'GO:1905524',
                 'GO:1904667', 'GO:1902499', 'GO:0043248', 'GO:0070628', 'GO:0031144', 'GO:0000502', 'GO:0031597', 'GO:0031595', 'GO:0005839',
                 'GO:0005838', 'GO:1903009', 'GO:0008537', 'GO:0022624', 'GO:0034515', 'GO:0080129', 'GO:0031603', 'GO:0031600', 'GO:0031601',
                 'GO:0031598', 'GO:0070682', 'GO:0090364', 'GO:0036402', 'GO:1902906', 'GO:1902907', 'GO:1990919', 'GO:1904855', 'GO:1904854',
                 'GO:0022623', 'GO:0008540', 'GO:0008541', 'GO:1990920', 'GO:0031615', 'GO:0031613', 'GO:0031612', 'GO:0031610', 'GO:1904327',
                 'GO:1990236', 'GO:0019773', 'GO:0019774', 'GO:0090363', 'GO:1902885', 'GO:0031604', 'GO:0031609', 'GO:0031606', 'GO:0031607',
                 'GO:0043161', 'GO:1902886', 'GO:1902887', 'GO:0039588', 'GO:0004175', 'GO:1990237', 'GO:0071630', 'GO:0071629', 'GO:0010498',
                 'GO:1904379', 'GO:0036369', 'GO:0004298', 'GO:1901483', 'GO:0006515', 'GO:1901799', 'GO:1901800', 'GO:1901484', 'GO:1901485',
                 'GO:0031593', 'GO:0071795', 'GO:0071796', 'GO:0070530', 'GO:0036435', 'GO:0071795', 'GO:0071796')

replication <- c('GO:0005657', 'GO:0006260', 'GO:0046809', 'GO:0006269', 'GO:0043596', 'GO:0006264', 'GO:0006270', 'GO:0006274', 'GO:0031297',
                 'GO:0000076', 'GO:0033260', 'GO:0033259', 'GO:0043111', 'GO:0098689', 'GO:0045004', 'GO:1990078', 'GO:0048478', 'GO:1902969',
                 'GO:0019045', 'GO:0071932', 'GO:0044786', 'GO:0031634', 'GO:0006261', 'GO:0003688', 'GO:0006275', 'GO:0031261', 'GO:0031298',
                 'GO:0097047', 'GO:0033314', 'GO:1902975', 'GO:1902979', 'GO:1902317', 'GO:0005662', 'GO:0005663', 'GO:0006336', 'GO:0006335',
                 'GO:0033567', 'GO:0090296', 'GO:0008156', 'GO:0045740', 'GO:2000621', 'GO:1990506', 'GO:1902294', 'GO:1902292', 'GO:1902973',
                 'GO:1902333', 'GO:0071170', 'GO:0071163', 'GO:0071946', 'GO:0036387', 'GO:0043598', 'GO:0043599', 'GO:0031582', 'GO:0005664',
                 'GO:0043599', 'GO:0006268', 'GO:0070517', 'GO:0090329', 'GO:0090592', 'GO:0051104', 'GO:1903459', 'GO:1903460', 'GO:1903466',
                 'GO:1903468', 'GO:0090001', 'GO:0097046', 'GO:0090298', 'GO:0090297', 'GO:0072438', 'GO:0043110', 'GO:0043137', 'GO:1902595',
                 'GO:1990505', 'GO:1990426', 'GO:1902297', 'GO:1902291', 'GO:1902971', 'GO:1902977', 'GO:1902315', 'GO:1902320', 'GO:0000809',
                 'GO:0032201', 'GO:0101017', 'GO:1990414', 'GO:0000808', 'GO:0030174', 'GO:0006271', 'GO:1903463', 'GO:1903467', 'GO:1903211',
                 'GO:0110035', 'GO:0033262', 'GO:0072437', 'GO:0072444', 'GO:0072441', 'GO:0098673', 'GO:0011000', 'GO:0045005', 'GO:2000105',
                 'GO:2000104', 'GO:1902596', 'GO:1902597', 'GO:1902298', 'GO:1902990', 'GO:1990943', 'GO:0101018', 'GO:0036388', 'GO:1903461',
                 'GO:1903465', 'GO:1903464', 'GO:0110025', 'GO:0072436', 'GO:0072443', 'GO:1902576', 'GO:1902681', 'GO:1902982', 'GO:1902983',
                 'GO:1904860', 'GO:0071171', 'GO:0010571', 'GO:0071807', 'GO:0032213', 'GO:0032297', 'GO:0032298', 'GO:0030894', 'GO:1903469',
                 'GO:1903221', 'GO:0072442', 'GO:1902296', 'GO:1902981', 'GO:0032214', 'GO:0032215', 'GO:0110026', 'GO:1902295', 'GO:1902299',
                 'GO:1902319', 'GO:0006267', 'GO:0110027', 'GO:1902318', 'GO:1990099', 'GO:1902561', 'GO:1902985')

folding <- c('GO:0006457', 'GO:0044183', 'GO:0006458', 'GO:0061077', 'GO:1903332', 'GO:1990727', 'GO:0051086', 'GO:0051084', 'GO:0051083',
             'GO:1903333', 'GO:1903334', 'GO:0007023', 'GO:0022417', 'GO:0034975', 'GO:1903644', 'GO:1990507', 'GO:0061992', 'GO:0060904',
             'GO:1903646', 'GO:1903645', 'GO:0051085', 'GO:0051087', 'GO:0101031', 'GO:0016531', 'GO:0033254', 'GO:1990565', 'GO:0034663',
             'GO:0016532', 'GO:0051131', 'GO:1902694', 'GO:1990507', 'GO:0061992', 'GO:0090034', 'GO:0072323', 'GO:1903645', 'GO:0090035',
             'GO:0051082')

chromatin <- c('GO:0000785', 'GO:0031497', 'GO:0031498', 'GO:0006325', 'GO:0006342', 'GO:0006338', 'GO:0030874', 'GO:0003682', 'GO:0070827',
               'GO:0001739', 'GO:0000789', 'GO:0000790', 'GO:0016569', 'GO:0031933', 'GO:0031490', 'GO:0005677', 'GO:0035327', 'GO:0035328',
               'GO:1990700', 'GO:0061793', 'GO:0006333', 'GO:0005521', 'GO:0061641', 'GO:0031935', 'GO:0090223', 'GO:0000183', 'GO:0099114',
               'GO:0006343', 'GO:0006344', 'GO:0006348', 'GO:0006346', 'GO:0030527', 'GO:1990280', 'GO:0031055', 'GO:0043035', 'GO:0043044',
               'GO:0035561', 'GO:0030702', 'GO:0061638', 'GO:1902275', 'GO:1990841', 'GO:0010848', 'GO:0010847', 'GO:0071168', 'GO:0090579',
               'GO:0031011', 'GO:0035101', 'GO:1905269', 'GO:1905268', 'GO:0048096', 'GO:0031936', 'GO:0031937', 'GO:0031048', 'GO:0035562',
               'GO:0035563', 'GO:1904834', 'GO:1990141', 'GO:0008623', 'GO:0033186', 'GO:0001672', 'GO:0090308', 'GO:0031938', 'GO:0090052',
               'GO:0061187', 'GO:1905634', 'GO:0070869', 'GO:0070868', 'GO:0070870', 'GO:0098578', 'GO:0035392', 'GO:0035390', 'GO:0071169',
               'GO:0097240', 'GO:0090309', 'GO:0031940', 'GO:0031939', 'GO:0090053', 'GO:0090310', 'GO:0061188', 'GO:0043156', 'GO:0045798',
               'GO:0045799', 'GO:2000749', 'GO:1904499', 'GO:0061644', 'GO:0120186', 'GO:0120187', 'GO:0034401', 'GO:0010964', 'GO:0060906',
               'GO:0001301', 'GO:1904500', 'GO:1904501', 'GO:0097549', 'GO:0016590', 'GO:0090115', 'GO:0070924', 'GO:0035391', 'GO:0001305',
               'GO:0001304', 'GO:0070923', 'GO:0070919', 'GO:0001308', 'GO:0070921', 'GO:0032116', 'GO:1904497', 'GO:1902368', 'GO:1902368',
               'GO:1902360', 'GO:1905635', 'GO:0071959', 'GO:0099404', 'GO:1905309', 'GO:0071922', 'GO:1905644', 'GO:2000715', 'GO:0000408',
               'GO:0071923', 'GO:2000716', 'GO:2000717', 'GO:0099403', 'GO:0071960', 'GO:0070603', 'GO:2000718', 'GO:1904907', 'GO:2000719',
               'GO:2000720', 'GO:1905646', 'GO:1905645', 'GO:1904908', 'GO:1904909', 'GO:0036414', 'GO:0016570', 'GO:0016571', 'GO:0016572',
               'GO:0016573', 'GO:0016574', 'GO:0016575', 'GO:0016576', 'GO:0016577', 'GO:0016578', 'GO:0043486', 'GO:0042393', 'GO:0106077',
               'GO:0001207', 'GO:0010390', 'GO:0071110', 'GO:0042054', 'GO:0035404', 'GO:0035405', 'GO:0035406', 'GO:0035035', 'GO:0000123', 
               'GO:0000412', 'GO:0061649', 'GO:0140372', 'GO:0043967', 'GO:0043966', 'GO:0043969', 'GO:0043968', 'GO:0070933', 'GO:0070932',
               'GO:0031493', 'GO:0004402', 'GO:0004407', 'GO:0000118', 'GO:1990258', 'GO:1990226', 'GO:1990164', 'GO:0033522', 'GO:0033523',
               'GO:0035064', 'GO:0106229', 'GO:0035363', 'GO:0106153', 'GO:0035097', 'GO:0035518', 'GO:0035521', 'GO:0070076', 'GO:0070077',
               'GO:0035173', 'GO:0106078', 'GO:0042826', 'GO:0061922', 'GO:0036205', 'GO:0032452', 'GO:0140069', 'GO:0140068', 'GO:0034969',
               'GO:0034968', 'GO:0006334', 'GO:0070577', 'GO:0018110', 'GO:0035400', 'GO:0035034', 'GO:0035184', 'GO:0035174', 'GO:0099077',
               'GO:0043989', 'GO:0043988', 'GO:0043981', 'GO:0043980', 'GO:0043983', 'GO:0043982', 'GO:0043985', 'GO:0043984', 'GO:0043987',
               'GO:0043990', 'GO:0043991', 'GO:0043998', 'GO:0043978', 'GO:0043977', 'GO:0043979', 'GO:0043970', 'GO:0043972', 'GO:0043971',
               'GO:0043974', 'GO:0043973', 'GO:0043976', 'GO:0043975', 'GO:0036123', 'GO:0036124', 'GO:0097043', 'GO:0070775', 'GO:1990245',
               'GO:1990259', 'GO:0080182', 'GO:0070734', 'GO:0070544', 'GO:0044648', 'GO:0031060', 'GO:0031063', 'GO:0031056', 'GO:0031059',
               'GO:0033182', 'GO:0033169', 'GO:0106185', 'GO:0035408', 'GO:0035409', 'GO:0035407', 'GO:0008334', 'GO:0035978', 'GO:0033100',
               'GO:0033127', 'GO:0098532', 'GO:0035033', 'GO:0035065', 'GO:0072370', 'GO:0072355', 'GO:0051567', 'GO:0051568', 'GO:1903584',
               'GO:0035522', 'GO:0035574', 'GO:0070078', 'GO:0070079', 'GO:0035267', 'GO:1902562', 'GO:0036410', 'GO:1990619', 'GO:1990596',
               'GO:0062072', 'GO:1990467', 'GO:1990468', 'GO:0061647', 'GO:0097692', 'GO:0071044', 'GO:0097676', 'GO:0061628', 'GO:1900049',
               'GO:0036413', 'GO:1990889', 'GO:0097198', 'GO:0036353', 'GO:0036351', 'GO:0036352', 'GO:1990678', 'GO:1990679', 'GO:0044013',
               'GO:0034729', 'GO:0034720', 'GO:0034773', 'GO:0034772', 'GO:0034771', 'GO:0034770', 'GO:0097725', 'GO:0044154', 'GO:0010452',
               'GO:0010484', 'GO:0010485', 'GO:0071572', 'GO:0071557', 'GO:0140373', 'GO:0034972', 'GO:0034971', 'GO:0034970', 'GO:0046811',
               'GO:0006398', 'GO:1990104', 'GO:0090239', 'GO:0043189', 'GO:0070776', 'GO:1901725', 'GO:0070537', 'GO:0070535', 'GO:0017136',
               'GO:0018024', 'GO:0031061', 'GO:0031064', 'GO:0031065', 'GO:0031062', 'GO:0031057', 'GO:0031058', 'GO:0033184', 'GO:0033183',
               'GO:0008469', 'GO:0045129', 'GO:0033129', 'GO:0033128', 'GO:0035066', 'GO:0035067', 'GO:1903586', 'GO:1903585', 'GO:0032777',
               'GO:0001208', 'GO:0036409', 'GO:0062060', 'GO:1990483', 'GO:0071045', 'GO:0071208', 'GO:0036206', 'GO:1990853', 'GO:2001166',
               'GO:1900050', 'GO:1900051', 'GO:0035979', 'GO:0035093', 'GO:0071204', 'GO:0043999', 'GO:0043992', 'GO:0043994', 'GO:0043993',
               'GO:0043996', 'GO:0043995', 'GO:0043997', 'GO:0061085', 'GO:0090241', 'GO:0090240', 'GO:0042800', 'GO:1990244', 'GO:1905471',
               'GO:1905435', 'GO:1990162', 'GO:1901727', 'GO:1901726', 'GO:0033749', 'GO:0033746', 'GO:0031151', 'GO:0070510', 'GO:0070611',
               'GO:0070612', 'GO:0031078', 'GO:0031066', 'GO:0032931', 'GO:0035401', 'GO:0035402', 'GO:0035403', 'GO:2000281', 'GO:0072354',
               'GO:0072371', 'GO:0051569', 'GO:0051570', 'GO:1901674', 'GO:0035575', 'GO:0035642', 'GO:0035175', 'GO:0042799', 'GO:0000414',
               'GO:0036408', 'GO:1902028', 'GO:0062122', 'GO:0062073', 'GO:1900112', 'GO:1900109', 'GO:0071207', 'GO:2001253', 'GO:1902464',
               'GO:0051864', 'GO:0036207', 'GO:0036208', 'GO:2000618', 'GO:2000615', 'GO:0044012', 'GO:0044014', 'GO:0044016', 'GO:0044015',
               'GO:0044018', 'GO:0044017', 'GO:0044019', 'GO:0044020', 'GO:0044023', 'GO:0044022', 'GO:0044025', 'GO:0044024', 'GO:0034739',
               'GO:0032129', 'GO:2001160', 'GO:2001167', 'GO:2001168', 'GO:0071440', 'GO:0071558', 'GO:0032454', 'GO:0032453', 'GO:0046974',
               'GO:0046975', 'GO:0046972', 'GO:0046976', 'GO:0034649', 'GO:0034648', 'GO:0034647', 'GO:0034721', 'GO:0061086', 'GO:0061087',
               'GO:1905473', 'GO:1905472', 'GO:1905437', 'GO:1905436', 'GO:0070511', 'GO:0070512', 'GO:0031068', 'GO:0031067', 'GO:1901314',
               'GO:0051574', 'GO:0051572', 'GO:0051573', 'GO:0051571', 'GO:1901675', 'GO:1901676', 'GO:0035616', 'GO:0000416', 'GO:0000415',
               'GO:2000620', 'GO:1902030', 'GO:1902029', 'GO:2001254', 'GO:2001255', 'GO:1900111', 'GO:1900110', 'GO:1900113', 'GO:1900114',
               'GO:0061866', 'GO:1902465', 'GO:1902466', 'GO:2000617', 'GO:2000619', 'GO:2000616', 'GO:1902649', 'GO:2001162', 'GO:2001161',
               'GO:0071442', 'GO:0071441', 'GO:0071894', 'GO:0035042', 'GO:1904173', 'GO:1901316', 'GO:1901315', 'GO:2000775', 'GO:0097372',
               'GO:0032041', 'GO:1902651', 'GO:1902650', 'GO:0046969', 'GO:0046970', 'GO:0032221', 'GO:0097044', 'GO:1904175', 'GO:1904174',
               'GO:2001173', 'GO:0033698', 'GO:0000125', 'GO:2000776', 'GO:2001174', 'GO:2001175', 'GO:0016581', 'GO:1990121', 'GO:2000751',
               'GO:2000873', 'GO:2000817', 'GO:0034080', 'GO:1903097', 'GO:0097549', 'GO:0034401', 'GO:1903098', 'GO:1903099')

GA <- c('GO:0010336', 'GO:0010331', 'GO:0010476', 'GO:0045487', 'GO:0009739', 'GO:0009686', 'GO:0009685', 'GO:0102713', 'GO:1905201', 'GO:0045544',
        'GO:0047927', 'GO:0033470', 'GO:0033469', 'GO:0102714', 'GO:0102716', 'GO:0102715', 'GO:0102712', 'GO:0102711', 'GO:0016707', 'GO:0102653',
        'GO:0102652', 'GO:0102663', 'GO:0102924', 'GO:0051779', 'GO:0102111', 'GO:0102119', 'GO:0102118', 'GO:0102117', 'GO:0102122', 'GO:0045543',
        'GO:0047928', 'GO:0071370', 'GO:0010341', 'GO:0010371', 'GO:0102738', 'GO:0102739', 'GO:0102123', 'GO:0102123', 'GO:0102125', 'GO:0102124',
        'GO:0010372', 'GO:0010373', 'GO:0103057', 'GO:0103056', 'GO:0052634', 'GO:0052635', 'GO:0103010', 'GO:0102972', 'GO:0103054', 'GO:0102755')

ABA <- c('GO:0080168', 'GO:1902265', 'GO:0010427', 'GO:0009687', 'GO:1902266', 'GO:0010294', 'GO:0010293', 'GO:0046345', 'GO:0009737', 'GO:0009688',
         'GO:0090440', 'GO:0010295', 'GO:0009724', 'GO:0009738', 'GO:0071215', 'GO:0010115', 'GO:0090359', 'GO:0051993', 'GO:1902418', 'GO:0009787',
         'GO:0010116', 'GO:1902417', 'GO:0009789', 'GO:0009788', 'GO:0075343', 'GO:1901527', 'GO:1990218')

splicing <- c('GO:0008380', 'GO:0034247', 'GO:0000394', 'GO:0000374', 'GO:0000373', 'GO:0000372', 'GO:1990935', 'GO:0006388', 'GO:0070054', 'GO:0016539',
              'GO:0043484', 'GO:0072669', 'GO:0000366', 'GO:0000398', 'GO:0033120', 'GO:0033119', 'GO:0045292', 'GO:0045291', 'GO:0002564', 'GO:0000380',
              'GO:0000365', 'GO:0000375', 'GO:0030909', 'GO:0048024', 'GO:0019801', 'GO:0019802', 'GO:0070274', 'GO:0000214', 'GO:0048026', 'GO:0048025',
              'GO:1905744', 'GO:0035048', 'GO:0000381', 'GO:1905746', 'GO:1905745', 'GO:0002563', 'GO:0071030', 'GO:0000376', 'GO:0000376', 'GO:0016607',
              'GO:0000213')

cell_wall <- c('GO:0005618', 'GO:0042545', 'GO:0042546', 'GO:0070726', 'GO:0052386', 'GO:0009530', 'GO:0009531', 'GO:0071555', 'GO:0044277',
               'GO:0009505', 'GO:0010384', 'GO:0071668', 'GO:0044347', 'GO:0099613', 'GO:0031506', 'GO:0031504', 'GO:0000032', 'GO:0000196',
               'GO:0016998', 'GO:0052543', 'GO:0052546', 'GO:0070592', 'GO:0005199', 'GO:0010339', 'GO:0010383', 'GO:0052325', 'GO:0009664',
               'GO:0044036', 'GO:0044038', 'GO:0009832', 'GO:0009828', 'GO:0009827', 'GO:0071554', 'GO:0009830', 'GO:0009829', 'GO:0034406',
               'GO:1990394', 'GO:2000652', 'GO:0052482', 'GO:0009834', 'GO:0009833', 'GO:0034410', 'GO:0044568', 'GO:0044567', 'GO:2000966',
               'GO:0010404', 'GO:0044348', 'GO:1903338', 'GO:0052541', 'GO:1901347', 'GO:1901348', 'GO:1902066', 'GO:0052324', 'GO:0071669',
               'GO:0010981', 'GO:2000967', 'GO:2000968', 'GO:1903339', 'GO:1903340', 'GO:0042547', 'GO:0052544', 'GO:0070597', 'GO:0070598',
               'GO:0060870', 'GO:1990076', 'GO:1902088', 'GO:2000939', 'GO:0090379', 'GO:1905588', 'GO:0080157', 'GO:2001009', 'GO:2000940',
               'GO:2000941', 'GO:1902089', 'GO:0009831', 'GO:0009920')

tubulin <- c('GO:0090042', 'GO:0015631', 'GO:0045298', 'GO:0042903', 'GO:0043015', 'GO:0043014', 'GO:0048487', 'GO:0000930', 'GO:0007021',
             'GO:0071929', 'GO:0070738', 'GO:0090043', 'GO:0004835', 'GO:0033566', 'GO:0019799', 'GO:0008275', 'GO:0008274', 'GO:0070463',
             'GO:1990727', 'GO:0000931', 'GO:1902481', 'GO:0034741', 'GO:0090044', 'GO:0070740', 'GO:0150067', 'GO:1904428', 'GO:0000927',
             'GO:0000924', 'GO:0007023', 'GO:0150068', 'GO:0150069', 'GO:0008568', 'GO:0055032', 'GO:0000928', 'GO:0055031', 'GO:0055033',
             'GO:0061494', 'GO:0061495', 'GO:1990735', 'GO:0110121', 'GO:0110120', 'GO:0005874', 'GO:0005828', 'GO:0005827', 'GO:0051013',
             'GO:0051012', 'GO:0015630', 'GO:0055028', 'GO:0008017', 'GO:0005880', 'GO:0005881', 'GO:0005876', 'GO:0005879', 'GO:0097427',
             'GO:1990752', 'GO:0007019', 'GO:0000235', 'GO:0007020', 'GO:0046785', 'GO:0034453', 'GO:0001578', 'GO:0099070', 'GO:0099071',
             'GO:0099609', 'GO:0099111', 'GO:0043622', 'GO:0031121', 'GO:0031122', 'GO:0005815', 'GO:0030981', 'GO:0030954', 'GO:0030953',
             'GO:0080175', 'GO:0060404', 'GO:0035371', 'GO:1905720', 'GO:0060172', 'GO:0005875', 'GO:1990498', 'GO:0003777', 'GO:0036449',
             'GO:0000226', 'GO:1990644', 'GO:0007018', 'GO:0007018', 'GO:0010938', 'GO:0010970', 'GO:0032932', 'GO:0099098', 'GO:0090222',
             'GO:0090064', 'GO:0099118', 'GO:0030473', 'GO:1990295', 'GO:0051011', 'GO:0051010', 'GO:0031113', 'GO:0031114', 'GO:0031109',
             'GO:0031023', 'GO:0031021', 'GO:0035372', 'GO:0047496', 'GO:1905759', 'GO:0098840', 'GO:1903754', 'GO:0106006', 'GO:1990537',
             'GO:1904511', 'GO:1904526', 'GO:0061673', 'GO:1990755', 'GO:0000923', 'GO:0061842', 'GO:0061863', 'GO:1990941', 'GO:0032117',
             'GO:0032118', 'GO:0034454', 'GO:0010968', 'GO:0036078', 'GO:0090221', 'GO:0031887', 'GO:0072384', 'GO:0036250', 'GO:0090221', 
             'GO:0090063', 'GO:0031534', 'GO:0031535', 'GO:0099112', 'GO:1904186', 'GO:1904185', 'GO:0031116', 'GO:0031117', 'GO:0031115',
             'GO:0070507', 'GO:0032886', 'GO:0031024', 'GO:0031025', 'GO:0070462', 'GO:1905833', 'GO:1905721', 'GO:1905725', 'GO:1905755',
             'GO:0072698', 'GO:0060632', 'GO:0072699', 'GO:1902513', 'GO:2000576', 'GO:2000575', 'GO:1904519', 'GO:0010005', 'GO:1902838',
             'GO:1902850', 'GO:1902816', 'GO:1990933', 'GO:1904825', 'GO:0007027', 'GO:0034643', 'GO:0034631', 'GO:1902619', 'GO:0140274',
             'GO:0034994', 'GO:0140210', 'GO:1905121', 'GO:0099606', 'GO:1903032', 'GO:1903033', 'GO:0099117', 'GO:1905509', 'GO:0030951',
             'GO:0031112', 'GO:0031111', 'GO:1903696', 'GO:0072385', 'GO:0072383', 'GO:0072382', 'GO:0072386', 'GO:0051415', 'GO:0075519',
             'GO:1901610', 'GO:1901609', 'GO:0098863', 'GO:1904518', 'GO:1902840', 'GO:1902839', 'GO:1902817', 'GO:1990810', 'GO:1904759',
             'GO:0034640', 'GO:1902620', 'GO:0140273', 'GO:1905185', 'GO:0090172', 'GO:0008569', 'GO:0008574', 'GO:0055031', 'GO:0055033',
             'GO:0106007', 'GO:1990734', 'GO:1990852', 'GO:0090176', 'GO:0090226', 'GO:0099110', 'GO:1903562', 'GO:1903563', 'GO:1990976',
             'GO:2000580', 'GO:2000577', 'GO:0140024', 'GO:0007099', 'GO:2000581', 'GO:2000582', 'GO:2000578', 'GO:2000579', 'GO:0008608',
             'GO:0051315', 'GO:0099607', 'GO:0051987', 'GO:0051986', 'GO:1902423', 'GO:1905115', 'GO:1902424', 'GO:1902425', 'GO:1905116',
             'GO:0051455', 'GO:1904967', 'GO:1904968', 'GO:0047497')

motility <- c('GO:0030286', 'GO:0070840', 'GO:0005868', 'GO:0045504', 'GO:0045505', 'GO:0045503', 'GO:0051959', 'GO:0003777', 'GO:2000574',
              'GO:2000576', 'GO:2000575', 'GO:0019894', 'GO:0005871', 'GO:0016938', 'GO:0016939', 'GO:0005873', 'GO:0005872', 'GO:2000577',
              'GO:2000578', 'GO:2000579', 'GO:2000580', 'GO:2000581', 'GO:2000582', 'GO:2000576', 'GO:2000575', 'GO:0008569', 'GO:0008574')

PH <- c('GO:0015979', 'GO:0009761', 'GO:0009760', 'GO:0010109', 'GO:0009765', 'GO:0019685', 'GO:0019684', 'GO:0019253', 'GO:1905156', 'GO:1905157',
        'GO:0009764', 'GO:0042548', 'GO:0010110', 'GO:0009763', 'GO:0009762', 'GO:0043155', 'GO:0009769', 'GO:0009768', 'GO:0080152', 'GO:0080153',
        'GO:0045156', 'GO:0045157', 'GO:0009579', 'GO:0042651', 'GO:0031976', 'GO:0031977', 'GO:0009534', 'GO:0009515', 'GO:0010547', 'GO:0055035',
        'GO:0031978', 'GO:0070116', 'GO:0010027', 'GO:0044160', 'GO:0009533', 'GO:0009535', 'GO:0009543', 'GO:0031676', 'GO:0031361', 'GO:0031360',
        'GO:0070117', 'GO:0070118', 'GO:0009503', 'GO:0010548', 'GO:0033654', 'GO:0035448', 'GO:0098807', 'GO:0030096', 'GO:0035449', 'GO:0098571',
        'GO:0098572', 'GO:0045038', 'GO:0033658', 'GO:0048493', 'GO:0035451', 'GO:0035450', 'GO:0019253', 'GO:0080152', 'GO:0080153', 'GO:0009854')

redox <- c('GO:0006979', 'GO:0070994', 'GO:0090403', 'GO:0034599', 'GO:0001306', 'GO:1902882', 'GO:1903201', 'GO:1903203', 'GO:0036475', 'GO:0036473',
           'GO:1902883', 'GO:1902884', 'GO:1900407', 'GO:0043556', 'GO:1903202', 'GO:1903204', 'GO:1903209', 'GO:1903223', 'GO:1900409', 'GO:1900408',
           'GO:0008631', 'GO:0032939', 'GO:0032938', 'GO:1902175', 'GO:1903376', 'GO:1902176', 'GO:1902177', 'GO:0036480', 'GO:2000815', 'GO:1903377',
           'GO:1903378', 'GO:0001324', 'GO:0001322', 'GO:0043619', 'GO:0036091', 'GO:0061417', 'GO:0097468', 'GO:0045730', 'GO:0080183', 'GO:0071455',
           'GO:0071456', 'GO:0001666', 'GO:0000302', 'GO:1901031', 'GO:0001315', 'GO:1903298', 'GO:1900038', 'GO:1900039', 'GO:1901033', 'GO:1901032')

respiration <- c('GO:0045333', 'GO:0009060', 'GO:0010230', 'GO:0043457', 'GO:1903715', 'GO:1901857', 'GO:1901856', 'GO:0008177', 'GO:0006099',
                 'GO:0019643', 'GO:0045246', 'GO:0006119', 'GO:0017077', 'GO:0002082', 'GO:0090324', 'GO:1903862', 'GO:2000275', 'GO:0018478',
                 'GO:0006121', 'GO:0050040', 'GO:1901857', 'GO:1901856', 'GO:1902957', 'GO:1902958', 'GO:0061621', 'GO:0061615', 'GO:0093001',
                 'GO:1904538', 'GO:0061616', 'GO:0006110', 'GO:0006096', 'GO:0045820', 'GO:0045821', 'GO:1904539', 'GO:1904540', 'GO:0006097',
                 'GO:2000874', 'GO:2000875', 'GO:2000876', 'GO:0050083', 'GO:0050441', 'GO:0004451', 'GO:0050083', 'GO:0004474', 'GO:0009514',
                 'GO:0031908', 'GO:0010111', 'GO:0046861', 'GO:0030061', 'GO:0005746', 'GO:0005750', 'GO:0005751', 'GO:0005747',
                 'GO:0005962', 'GO:0000276', 'GO:0000275', 'GO:0000274', 'GO:0062011', 'GO:0005756', 'GO:0042652', 'GO:0042653')

seed_develoment <- c('GO:0010162', 'GO:0080112', 'GO:1990068', 'GO:0097548', 'GO:0048317', 'GO:0048316', 'GO:0010431',
                    'GO:0090377', 'GO:0090378', 'GO:0090376', 'GO:0090380', 'GO:0010214', 'GO:0010344', 'GO:0080113', 'GO:2000033',
                    'GO:1990137', 'GO:0080050', 'GO:2000034', 'GO:2000692', 'GO:2000693', 'GO:0010187', 'GO:1902039', 'GO:1902040',
                    'GO:0090379')

?embryogenesis <- c('GO:0009790', 'GO:0048314', 'GO:0009553', 'GO:0009558', 'GO:0009562', 'GO:0010072', 'GO:0009559', 'GO:0009560',
                   'GO:0043045', 'GO:0023002', 'GO:0010069', 'GO:0045691', 'GO:0045694', 'GO:0023003', 'GO:0045695', 'GO:0045692',
                   'GO:0045693', 'GO:0045696', 'GO:0009561', 'GO:0010262', 'GO:0048598', 'GO:0090421', 'GO:0048508', 'GO:0010086',
                   'GO:0010064')

seed_germination <- c('GO:0009845', 'GO:0010029', 'GO:0010231', 'GO:0080001', 'GO:0010030', 'GO:0048623', 'GO:0048838', 'GO:1902039',
                      'GO:0010187', 'GO:0048700', 'GO:0009793', 'GO:0098755', 'GO:0048354', 'GO:0048359', 'GO:0090351', 'GO:1900140',
                      'GO:0009880')

meristems <- c('GO:0010072', 'GO:0090421', 'GO:0048508', 'GO:0035266', 'GO:0010022', 'GO:0010014', 'GO:0010073', 'GO:0048507',
               'GO:0010065', 'GO:0090421', 'GO:0010066', 'GO:0010071', 'GO:0048508', 'GO:0010451', 'GO:0010450', 'GO:0010449',
               'GO:0010448', 'GO:0010582', 'GO:0009933', 'GO:0090506', 'GO:0010075', 'GO:1902182', 'GO:0010074', 'GO:0048509',
               'GO:0010072', 'GO:0010082', 'GO:0010081', 'GO:0010083', 'GO:0010080', 'GO:0010077', 'GO:0010076', 'GO:0010079',
               'GO:0010078', 'GO:0009934', 'GO:1902183', 'GO:0010492', 'GO:0007639', 'GO:1902185', 'GO:1902184', 'GO:0010228',
               'GO:0010199', 'GO:0010443', 'GO:0048506')

transposons <- c('GO:0070893', 'GO:0070894', 'GO:0070896', 'GO:0070895', 'GO:0120085', 'GO:0003964',
                 'GO:0015074', 'GO:0009294')

```

There are also cases when the full-transcriptome gene universe size is undesirable for the over-representation tests. These include annotations of subsets derived from a set of transcripts referring to a specific functional category, such as overexpression spots on the category-wise self-organizing maps. The following function produces a universe subset fit for these functions:

``` {r GO_universe_function, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE, error=FALSE}

create_GO_geneUniverse <- function(annotation, pattern){
  full_mapping <- annotation %>% filter(grepl(paste0('(', paste(pattern, collapse = ')|('), ')', sep = ''), GO_BLAST)) %>% dplyr::select(transcript_id, GO_BLAST) %>% mutate(transcript_id = as.character(transcript_id), GO_BLAST = as.character(GO_BLAST) %>% sapply(function(x) strsplit(x, ',')))
  names(full_mapping$GO_BLAST) <- full_mapping$transcript_id
full_mapping <- full_mapping$GO_BLAST

  return(full_mapping)
}

full_mapping_ext <- create_GO_geneUniverse(annot_full_ext, '*')
geneUniverse_ext <- names(full_mapping_ext)

```

Though no confidence intervals could be plotted in this case, barplots seem to be a convenient way to visualize over-represented terms. The following function combines top terms from both up- and downregulated transcript subsets and plots them colored respective to regulation direction: 

```{r GO barplots}

GO_barplotter <- function(upregulated, downregulated, top_features = 10) {
  
  if (!is.na(upregulated)) {
    a <- min(top_features, nrow(upregulated))
  upregulated <- upregulated[1:a,]
  } else {
    a <- 1
  }
  
  if (!is.na(downregulated)) {
    b <- min(top_features, nrow(downregulated))
    downregulated <- downregulated[1:b,]
  } else {
    b <- 1
  }
  
  combined_dataset <- rbind(upregulated, downregulated) %>% mutate(condition = c(rep('Upregulated', a), rep('Downregulated', b))) %>% arrange(desc(Percentage_of_Differentially_Expressed_Genes_from_Number_Of_Annotated_Genes)) %>% mutate(Term = factor(Term, levels = unique(Term))) %>% filter_all(all_vars(!is.na(.)))
  ggplot(combined_dataset, aes(y=Percentage_of_Differentially_Expressed_Genes_from_Number_Of_Annotated_Genes,
                                                               x=reorder(Term, Percentage_of_Differentially_Expressed_Genes_from_Number_Of_Annotated_Genes, max), fill = condition)) + 
    geom_bar(stat = 'identity', position = 'identity', width = 0.8) + coord_flip() +
    theme(text = element_text(size=24, family = 'arial', color='black'), axis.text.x = element_text(color = 'black', family = 'arial'),
          axis.title.x = element_text(color = 'black', family = 'arial', size = 20),
          axis.text.y = element_text(color = 'black', family = 'arial', hjust = 1),
          axis.title.y = element_blank(),
          axis.line.x = element_line(color = 'black'),
          axis.ticks.length.y =  unit(0, 'lines'),
          panel.background = element_blank(), legend.text = element_text(size = 22), legend.title = 
            element_text(size = 28)) + 
    labs(y = '% of annotated genes') + 
    scale_fill_manual('condition', values = c('Upregulated' = '#068013', 'Downregulated' = '#9c1718'))
  }


```

## Transcription Factor MapMan Annotation
One of the best things about MapMan annotation is the readability of its terms and their ease of interpretation. Since we have a special interest in differential expression of transcription factors, we introduce a pattern covering manually picked MapMan TF terms that we will use further.

```{r MapMan_TFs, echo=TRUE, eval = TRUE, warning=FALSE, error=FALSE, message=FALSE}

tf_template <- '(GeBP like)|(\\.ARR)|(CCAAT box binding factor family)|(C2C2\\(Zn\\) YABBY family)|(C2C2\\(Zn\\) DOF zinc finger family)|(ovate family OFP)|
                           (BBR/BPC)|(plant TF \\(pbf2\\))|(BSD domain containing family)|(AP2/EREBP, APETALA2/Ethylene-responsive element binding protein family)|
                           (zf-HD)|(sigma like plant)|(ARF, Auxin Response Factor family)|(ARF, Auxin Response Factor family)|(C2C2\\(Zn\\) CO-like, Constans-like zinc finger family)|
                           (CCAAT box binding factor family, HAP2)|(NIN-like bZIP-related family)|(GRF zinc finger family)|(Alfin-like)|(ELF3)|(transcription factor)'

TF_tracker <- function(DE_df) {
  TF_df <- annot_full_ext[grepl('regulation of transcription', annot_full_ext$MapMan_term),] %>% filter(transcript_id %in% DE_df$target_id) %>% dplyr::select(transcript_id, MapMan_term) %>% mutate(MapMan_term = MapMan_term %>% sapply(function(x) unlist(strsplit(x, ';'))[grepl('regulation of transcription', unlist(strsplit(x, ';')))])) %>% mutate(TF_Class = MapMan_term %>% sapply(function(x) x %>% strsplit(.,'\\.') %>% unlist() %>% unname() %>% nth(3)))

TF_df <- table(unlist(TF_df$TF_Class)) %>% as.data.frame() %>% arrange(desc(Freq))
 ony 
assign(paste('TF', deparse(substitute(DE_df)), sep = '_'), TF_df, envir = .GlobalEnv)
  

}

```



## Differential Expression Analysis



### Sprint-2 - native assembly

```{r Sprint2_DE, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
sample_path_Sprint2 <- file.path('kallisto_Sprint2', dir(file.path('kallisto_Sprint2'))[1:8])
metadata_Sprint2 <- read.table(file.path('kallisto_Sprint2', 'metadata.txt'), header = T, stringsAsFactors=F) %>% mutate(path = sample_path_Sprint2, DAP = as.factor(DAP))
so_Sprint2 <- sleuth_prep(metadata_Sprint2, extra_bootstrap_summary = T)
so_Sprint2 <- sleuth_fit(so_Sprint2, ~1, 'reduced')
so_Sprint2 <- sleuth_fit(so_Sprint2, ~DAP, 'full')
so_Sprint2 <- sleuth_lrt(so_Sprint2, 'reduced', 'full')
so_Sprint2 <- sleuth_wt(so_Sprint2, 'DAP20')

lrt_Sprint2 <- sleuth_results(so_Sprint2, 'reduced:full', 'lrt', show_all = F)
wt_Sprint2 <- sleuth_results(so_Sprint2, 'DAP20', 'wt', show_all = F)

joint_results_orig <- wt_Sprint2 %>% filter(target_id %in% lrt_Sprint2$target_id)
joint_results <- wt_Sprint2 %>% filter(qval < 0.001, target_id %in% lrt_Sprint2[lrt_Sprint2$qval < 0.001, 'target_id'])
up_Sprint2 <- joint_results %>% filter(b > 0.5)
down_Sprint2 <- joint_results %>% filter(b < -0.5)


```


### Sprint-2 - сombined assembly

```{r Sprint2_DE_comb, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

sample_path_Sprint2_com <- file.path('kallisto_outputs_combined_assembly', dir(file.path('kallisto_outputs_combined_assembly'))[1:8])
metadata_Sprint2_com <- read.table(file.path('kallisto_Sprint2', 'metadata.txt'), header = T, stringsAsFactors=F) %>% mutate(path = sample_path_Sprint2_com, DAP = as.factor(DAP))

cso_Sprint2 <- sleuth_prep(metadata_Sprint2_com, extra_bootstrap_summary = T)
cso_Sprint2 <- sleuth_fit(cso_Sprint2, ~1, 'reduced')
cso_Sprint2 <- sleuth_fit(cso_Sprint2, ~DAP, 'full')
cso_Sprint2 <- sleuth_lrt(cso_Sprint2, 'reduced', 'full')
cso_Sprint2 <- sleuth_wt(cso_Sprint2, 'DAP20')

lrt_Sprint2_com <- sleuth_results(cso_Sprint2, 'reduced:full', 'lrt', show_all = F)
wt_Sprint2_com <- sleuth_results(cso_Sprint2, 'DAP20', 'wt', show_all = F)

joint_results_com_orig <- wt_Sprint2_com %>% filter(target_id %in% lrt_Sprint2_com$target_id)
joint_results_com <- wt_Sprint2_com %>% filter(qval < 0.001, target_id %in% lrt_Sprint2_com[lrt_Sprint2_com$qval < 0.001, 'target_id'])
up_Sprint2_com <- joint_results_com %>% filter(b > 0.5)
down_Sprint2_com <- joint_results_com %>% filter(b < -0.5)

```

### Zhewan-1

```{r Zhewan1_DE, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

sample_path_Zhewan1 <- file.path('kallisto_outputs_combined_assembly', dir(file.path('kallisto_outputs_combined_assembly'))[10:17])
metadata_Zhewan1 <- read.table(file.path('kallisto_subsamples_external', 'Zhewan_metadata.txt'), header = T, stringsAsFactors=F) %>% mutate(path = sample_path_Zhewan1, DAP = as.factor(DAP))

so_Zhewan1 <- sleuth_prep(metadata_Zhewan1, extra_bootstrap_summary = T)
so_Zhewan1 <- sleuth_fit(so_Zhewan1, ~1, 'reduced')
so_Zhewan1 <- sleuth_fit(so_Zhewan1, ~DAP, 'full')
so_Zhewan1 <- sleuth_lrt(so_Zhewan1, 'reduced', 'full')
so_Zhewan1 <- sleuth_wt(so_Zhewan1, 'DAP25')

lrt_Zhewan1 <- sleuth_results(so_Zhewan1, 'reduced:full', 'lrt', show_all = F)
wt_Zhewan1 <- sleuth_results(so_Zhewan1, 'DAP25', 'wt', show_all = F)

joint_results_Zhewan1_orig <- wt_Zhewan1 %>% filter(target_id %in% lrt_Zhewan1$target_id)
joint_results_Zhewan1 <- wt_Zhewan1 %>% filter(qval < 0.001, target_id %in% lrt_Zhewan1[lrt_Zhewan1$qval < 0.001, 'target_id'])
up_Zhewan1 <- joint_results_Zhewan1 %>% filter(b > 0.5)
down_Zhewan1 <- joint_results_Zhewan1 %>% filter(b < -0.5)

```

### Zhongwan_6

```{r Zhongwan6_DE, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
sample_path_Zhongwan6 <- file.path('kallisto_outputs_combined_assembly', dir(file.path('kallisto_outputs_combined_assembly'))[18:25])
metadata_Zhongwan6 <- read.table(file.path('kallisto_subsamples_external', 'Zhongwan_metadata.txt'), header = T, stringsAsFactors=F) %>% mutate(path = sample_path_Zhongwan6, DAP = as.factor(DAP))

so_Zhongwan6 <- sleuth_prep(metadata_Zhongwan6, extra_bootstrap_summary = T)
so_Zhongwan6 <- sleuth_fit(so_Zhongwan6, ~1, 'reduced')
so_Zhongwan6 <- sleuth_fit(so_Zhongwan6, ~DAP, 'full')
so_Zhongwan6 <- sleuth_lrt(so_Zhongwan6, 'reduced', 'full')
so_Zhongwan6 <- sleuth_wt(so_Zhongwan6, 'DAP25')

lrt_Zhongwan6 <- sleuth_results(so_Zhongwan6, 'reduced:full', 'lrt', show_all = F)
wt_Zhongwan6 <- sleuth_results(so_Zhongwan6, 'DAP25', 'wt', show_all = F)

joint_results_Zhongwan6_orig <- wt_Zhongwan6 %>% filter(target_id %in% lrt_Zhongwan6$target_id)
joint_results_Zhongwan6 <- wt_Zhongwan6 %>% filter(qval < 0.001, target_id %in% lrt_Zhongwan6[lrt_Zhongwan6$qval < 0.001, 'target_id'])
up_Zhongwan6 <- joint_results_Zhongwan6 %>% filter(b > 0.5)
down_Zhongwan6 <- joint_results_Zhongwan6 %>% filter(b < -0.5)

```


### 10 DAP: Zhewan-1 vs Sprint-2

```{r 10DAP_Sprint_Zhewan, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
metadata_10SpriZhe <- rbind(metadata_Sprint2_com, metadata_Zhewan1) %>% filter(DAP == '10') %>% mutate(cultivar = c(rep('Sprint-2', 4), rep('Zhewan-1', 4)))

so_10SpriZhe <- sleuth_prep(metadata_10SpriZhe, extra_bootstrap_summary = T)
so_10SpriZhe <- sleuth_fit(so_10SpriZhe, ~1, 'reduced')
so_10SpriZhe <- sleuth_fit(so_10SpriZhe, ~cultivar, 'full')
so_10SpriZhe <- sleuth_lrt(so_10SpriZhe, 'reduced', 'full')
so_10SpriZhe <- sleuth_wt(so_10SpriZhe, 'cultivarZhewan-1')

lrt_10SpriZhe <- sleuth_results(so_10SpriZhe, 'reduced:full', 'lrt', show_all = F)
wt_10SpriZhe <- sleuth_results(so_10SpriZhe, 'cultivarZhewan-1', 'wt', show_all = F)

joint_results_10SpriZhe_orig <- wt_10SpriZhe %>% filter(target_id %in% lrt_10SpriZhe$target_id)
joint_results_10SpriZhe <- wt_10SpriZhe %>% filter(qval < 0.001, target_id %in% lrt_10SpriZhe[lrt_10SpriZhe$qval < 0.001, 'target_id'])
up_10SpriZhe <- joint_results_10SpriZhe %>% filter(b > 0.5)
down_10SpriZhe <- joint_results_10SpriZhe %>% filter(b < -0.5)

```


### 10 DAP: Zhongwan-6 vs Sprint-2 

```{r 10DAP_Sprint_Zhongwan, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
metadata_10SpriZho <- rbind(metadata_Sprint2_com, metadata_Zhongwan6) %>% filter(DAP == '10') %>% mutate(cultivar = c(rep('Sprint-2', 4), rep('Zhongwan-6', 4)))

so_10SpriZho <- sleuth_prep(metadata_10SpriZho, extra_bootstrap_summary = T)
so_10SpriZho <- sleuth_fit(so_10SpriZho, ~1, 'reduced')
so_10SpriZho <- sleuth_fit(so_10SpriZho, ~cultivar, 'full')
so_10SpriZho <- sleuth_lrt(so_10SpriZho, 'reduced', 'full')
so_10SpriZho <- sleuth_wt(so_10SpriZho, 'cultivarZhongwan-6')

lrt_10SpriZho <- sleuth_results(so_10SpriZho, 'reduced:full', 'lrt', show_all = F)
wt_10SpriZho <- sleuth_results(so_10SpriZho, 'cultivarZhongwan-6', 'wt', show_all = F)

joint_results_10SpriZho_orig <- wt_10SpriZho %>% filter(target_id %in% lrt_10SpriZho$target_id)
joint_results_10SpriZho <- wt_10SpriZho %>% filter(qval < 0.001, target_id %in% lrt_10SpriZho[lrt_10SpriZho$qval < 0.001, 'target_id'])
up_10SpriZho <- joint_results_10SpriZho %>% filter(b > 0.5)
down_10SpriZho <- joint_results_10SpriZho %>% filter(b < -0.5)

```

### 20-25 DAP: Zhewan-1 vs Sprint-2

```{r 20DAP_Sprint_Zhewan, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
metadata_20SpriZhe <- rbind(metadata_Sprint2_com, metadata_Zhewan1) %>% filter(DAP %in% c('20', '25')) %>% mutate(cultivar = c(rep('Sprint-2', 4), rep('Zhewan-1', 4)))

so_20SpriZhe <- sleuth_prep(metadata_20SpriZhe, extra_bootstrap_summary = T)
so_20SpriZhe <- sleuth_fit(so_20SpriZhe, ~1, 'reduced')
so_20SpriZhe <- sleuth_fit(so_20SpriZhe, ~cultivar, 'full')
so_20SpriZhe <- sleuth_lrt(so_20SpriZhe, 'reduced', 'full')
so_20SpriZhe <- sleuth_wt(so_20SpriZhe, 'cultivarZhewan-1')

lrt_20SpriZhe <- sleuth_results(so_20SpriZhe, 'reduced:full', 'lrt', show_all = F)
wt_20SpriZhe <- sleuth_results(so_20SpriZhe, 'cultivarZhewan-1', 'wt', show_all = F)

joint_results_20SpriZhe_orig <- wt_20SpriZhe %>% filter(target_id %in% lrt_20SpriZhe$target_id)
joint_results_20SpriZhe <- wt_20SpriZhe %>% filter(qval < 0.001, target_id %in% lrt_20SpriZhe[lrt_20SpriZhe$qval < 0.001, 'target_id'])
up_20SpriZhe <- joint_results_20SpriZhe %>% filter(b > 0.5)
down_20SpriZhe <- joint_results_20SpriZhe %>% filter(b < -0.5)

```

### 20-25 DAP: Zhongwan-6 vs Sprint-2

```{r 20DAP_Sprint_Zhongwan, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}
metadata_20SpriZho <- rbind(metadata_Sprint2_com, metadata_Zhongwan6) %>% filter(DAP %in% c('20', '25')) %>% mutate(cultivar = c(rep('Sprint-2', 4), rep('Zhongwan-6', 4)))

so_20SpriZho <- sleuth_prep(metadata_20SpriZho, extra_bootstrap_summary = T)
so_20SpriZho <- sleuth_fit(so_20SpriZho, ~1, 'reduced')
so_20SpriZho <- sleuth_fit(so_20SpriZho, ~cultivar, 'full')
so_20SpriZho <- sleuth_lrt(so_20SpriZho, 'reduced', 'full')
so_20SpriZho <- sleuth_wt(so_20SpriZho, 'cultivarZhongwan-6')

lrt_20SpriZho <- sleuth_results(so_20SpriZho, 'reduced:full', 'lrt', show_all = F)
wt_20SpriZho <- sleuth_results(so_20SpriZho, 'cultivarZhongwan-6', 'wt', show_all = F)

joint_results_20SpriZho_orig <- wt_20SpriZho %>% filter(target_id %in% lrt_20SpriZho$target_id)
joint_results_20SpriZho <- wt_20SpriZho %>% filter(qval < 0.001, target_id %in% lrt_20SpriZho[lrt_20SpriZho$qval < 0.001, 'target_id'])
up_20SpriZho <- joint_results_20SpriZho %>% filter(b > 0.5)
down_20SpriZho <- joint_results_20SpriZho %>% filter(b < -0.5)

```

### Subset and intersection sizes

```{r subset_sizes}

DE_numbers <- data.frame('direction' = c('up', 'down'), 'number of DEGs' = c(nrow(up_Sprint2), nrow(down_Sprint2)))

tiff(filename = file.path('revision_round1_supplementaries', 'DE_upsets_and_bars', 'Sprint2_DEGs.tiff'),
     width = 16, height = 20, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

ggplot(DE_numbers, aes(x=direction, y=number.of.DEGs, fill = direction)) + 
  geom_bar(stat = 'identity', position = 'identity', width = 0.5) + 
  theme_bw() + theme(panel.grid.major.x = element_blank(),
                     axis.title.x = element_text(size = 20), axis.title.y = element_text(size = 20),
                     axis.text = element_text(size = 18), legend.position = 'none') + labs(x='Direction of DE', y = 'Number of DEGs') +
  scale_fill_manual('direction', values = c('up' = '#068013', 'down' = '#9c1718'))

dev.off()



populate_venn_df <- function(subset) {
  checker <- function(num, sub) {
    if(annot_full_ext$transcript_id[num] %in% sub[,1]) return(1)
    else return(0)
  }
  DE_values <- sapply(1:nrow(annot_full_ext), function(x) checker(x, subset))
return(DE_values)
  }

venn_df <- data.frame('Sprint-2(10) vs Sprint-2(20), up' = populate_venn_df(up_Sprint2_com),
                      'Sprint-2(10) vs Sprint-2(20), down' = populate_venn_df(down_Sprint2_com),
                      'Sprint-2(10) vs Zhewan-1(10), up' = populate_venn_df(up_10SpriZhe),
                      'Sprint-2(10) vs Zhewan-1(10), down' = populate_venn_df(down_10SpriZhe),
                      'Sprint-2(10) vs Zhongwan-6(10), up' = populate_venn_df(up_10SpriZho),
                      'Sprint-2(10) vs Zhongwan-6(10), down' = populate_venn_df(down_10SpriZho),
                      'Sprint-2(20) vs Zhewan-1(25), up' = populate_venn_df(up_20SpriZhe),
                      'Sprint-2(20) vs Zhewan-1(25), down' = populate_venn_df(down_20SpriZhe),
                      'Sprint-2(20) vs Zhongwan-6(25), up' = populate_venn_df(up_20SpriZho),
                      'Sprint-2(20) vs Zhongwan-6(25), down' = populate_venn_df(down_20SpriZho)) %>% set_colnames(c(
                        'Sprint-2(10) vs Sprint-2(20), up', 'Sprint-2(10) vs Sprint-2(20), down',
                        'Sprint-2(10) vs Zhewan-1(10), up', 'Sprint-2(10) vs Zhewan-1(10), down',
                        'Sprint-2(10) vs Zhongwan-6(10), up', 'Sprint-2(10) vs Zhongwan-6(10), down',
                        'Sprint-2(20) vs Zhewan-1(25), up', 'Sprint-2(20) vs Zhewan-1(25), down',
                        'Sprint-2(20) vs Zhongwan-6(25), up', 'Sprint-2(20) vs Zhongwan-6(25), down'
                      )) %>% set_colnames(sapply(colnames(.), function(x) gsub('(,)? ', '_', x, perl = T))) %>% 
  set_colnames(sapply(colnames(.), function(x) gsub('\\(', '_', x))) %>% set_colnames(sapply(colnames(.), function(x) gsub('\\)', '', x))) %>% set_colnames(sapply(colnames(.), function(x) gsub('-', '', x)))

tiff(filename = file.path('revision_round1_supplementaries', 'DE_upsets_and_bars', 'Figure3.tiff'),
      width = 44, height = 24, units = "cm", pointsize = 12,
      compression = "lzw",
      bg = "white", res = 520, family = "",
      type = "cairo")

upset(venn_df, sets = colnames(venn_df), sets.bar.color = '#6dc585',
      order.by = 'freq', text.scale = c(2.1, 2, 1.8, 1.5, 2, 2), mainbar.y.label = 'Number of DEGs',
      intersections = list(list('Sprint2_10_vs_Zhongwan6_10_down'), list('Sprint2_10_vs_Zhewan1_10_down'),
                           list('Sprint2_10_vs_Zhongwan6_10_up'), list('Sprint2_10_vs_Zhewan1_10_up'),
                           list('Sprint2_20_vs_Zhongwan6_25_down'), list('Sprint2_20_vs_Zhewan1_25_down'),
                           list('Sprint2_20_vs_Zhongwan6_25_up'), list('Sprint2_20_vs_Zhewan1_25_up'),
                           list('Sprint2_10_vs_Zhongwan6_10_down', 'Sprint2_10_vs_Zhewan1_10_down', 'Sprint2_10_vs_Sprint2_20_down'),
                           list('Sprint2_10_vs_Zhongwan6_10_down', 'Sprint2_10_vs_Zhewan1_10_down'),
                           list('Sprint2_10_vs_Zhongwan6_10_up', 'Sprint2_10_vs_Zhewan1_10_up', 'Sprint2_10_vs_Sprint2_20_up'),
                           list('Sprint2_10_vs_Zhongwan6_10_up', 'Sprint2_10_vs_Sprint2_20_up'),
                           list('Sprint2_10_vs_Zhewan1_10_up', 'Sprint2_10_vs_Sprint2_20_up'),
                           list('Sprint2_20_vs_Zhewan1_25_down', 'Sprint2_20_vs_Zhongwan6_25_down', 'Sprint2_10_vs_Zhewan1_10_down', 'Sprint2_10_vs_Zhongwan6_10_down'),
                           list('Sprint2_20_vs_Zhewan1_25_down', 'Sprint2_20_vs_Zhongwan6_25_down', 'Sprint2_10_vs_Sprint2_20_up'),
                           list('Sprint2_20_vs_Zhewan1_25_down', 'Sprint2_20_vs_Zhongwan6_25_down'),
                           list('Sprint2_10_vs_Zhongwan6_10_up', 'Sprint2_10_vs_Zhewan1_10_up'),
                           list('Sprint2_10_vs_Sprint2_20_down', 'Sprint2_20_vs_Zhewan1_25_down'),
                           list('Sprint2_10_vs_Sprint2_20_down', 'Sprint2_20_vs_Zhewan1_25_up'),
                           list('Sprint2_10_vs_Zhongwan6_10_down', 'Sprint2_20_vs_Zhongwan6_25_down'),
                           list('Sprint2_10_vs_Zhongwan6_10_down', 'Sprint2_20_vs_Zhongwan6_25_up')))
dev.off()

tiff(filename = file.path('revision_round1_supplementaries', 'DE_upsets_and_bars', 'FigureS1'),
      width = 64, height = 24, units = "cm", pointsize = 12,
      compression = "lzw",
      bg = "white", res = 520, family = "",
      type = "cairo")

upset(venn_df, sets = colnames(venn_df), sets.bar.color = '#6dc585',
      order.by = 'freq', text.scale = c(2.1, 2, 1.8, 1.5, 2, 2), mainbar.y.label = 'Number of DEGs')

dev.off()


SdZhedZhod <- annot_full_ext %>% filter(transcript_id %in% down_Sprint2_com$target_id & transcript_id %in% down_10SpriZhe$target_id & transcript_id %in% down_10SpriZho$target_id) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)

# these stand for all commonly downregulated transcripts in Zhewan-1 and Zhongwan-6

ZhedZhod <- annot_full_ext %>% filter(transcript_id %in% down_10SpriZhe$target_id & transcript_id %in% down_10SpriZho$target_id) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)

ZheuZhou <- annot_full_ext %>% filter(transcript_id %in% up_10SpriZhe$target_id & transcript_id %in% up_10SpriZho$target_id) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)

SuZheuZhou <- annot_full_ext %>% filter(transcript_id %in% up_10SpriZhe$target_id & transcript_id %in% up_10SpriZho$target_id & transcript_id %in% up_Sprint2_com$target_id) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)

ZhedZhod20 <- annot_full_ext %>% filter(transcript_id %in% down_20SpriZhe$target_id & transcript_id %in% down_20SpriZho$target_id) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)
SuZhedZhod20 <- annot_full_ext %>% filter(transcript_id %in% down_20SpriZhe$target_id & transcript_id %in% down_20SpriZho$target_id & transcript_id %in% up_Sprint2_com$target_id) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)
ZheuZhou_20 <- annot_full_ext %>% filter(transcript_id %in% up_20SpriZhe$target_id & transcript_id %in% up_20SpriZho$target_id) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)

ZhedZhod_ultimate <- annot_full_ext %>% filter(transcript_id %in% down_20SpriZhe$target_id & transcript_id %in% down_20SpriZho$target_id & transcript_id %in% down_10SpriZhe$target_id & transcript_id %in% down_10SpriZho$target_id & !(transcript_id %in% up_Sprint2_com$target_id) & !(transcript_id %in% down_Sprint2_com$target_id)) %>% dplyr::rename(target_id = transcript_id) %>% dplyr::select(target_id)

sapply(aspects, function(x) iterating_function_GO(SdZhedZhod, x, 0.01, full_mapping, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(ZhedZhod, x, 0.01, full_mapping, geneUniverse_ext))
BP_SdZhedZhod <- BP_SdZhedZhod[-c(16),]
sapply(aspects, function(x) iterating_function_GO(ZheuZhou, x, 0.01, full_mapping, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(SuZheuZhou, x, 0.01, full_mapping, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(ZhedZhod20, x, 0.01, full_mapping, geneUniverse_ext))
BP_ZhedZhod20 <- BP_ZhedZhod20[-c(14),]
sapply(aspects, function(x) iterating_function_GO(SuZhedZhod20, x, 0.01, full_mapping, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(ZheuZhou_20, x, 0.01, full_mapping, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(ZhedZhod_ultimate, x, 0.01, full_mapping, geneUniverse_ext))

sapply(ls()[grepl('.*SdZhedZhod', ls())], function(x) x %>% get() %>% write.table(., file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'SdZhedZhod',paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))

write.table(ZhedZhod, file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod', 'ZhedZhod.tsv'), sep = '\t', col.names = T, row.names = F)
sapply(ls()[grepl('((BP)|(CC)|(MF))_ZhedZhod', ls())], function(x) x %>% get() %>% write.table(., file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod',paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))

sapply(ls()[grepl('ZhedZhod20', ls())], function(x) x %>% get() %>% write.table(., file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod20',paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))

sapply(ls()[grepl('^(((BP)|(CC)|(MF))_)?ZheuZhou$', ls())], function(x) x %>% get() %>% write.table(., file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou',paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))
sapply(ls()[grepl('^(((BP)|(CC)|(MF))_)?ZheuZhou_', ls())], function(x) x %>% get() %>% write.table(., file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou_20',paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))

```


### GO Over-representation of DEG sets

```{r GO_OR, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE, results='hide'}

sapply(aspects, function(x) iterating_function_GO(up_Sprint2, x, 0.01, full_mapping, geneUniverse))
sapply(aspects, function(x) iterating_function_GO(down_Sprint2, x, 0.01, full_mapping, geneUniverse))
sapply(aspects, function(x) iterating_function_GO(up_Sprint2_com, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(down_Sprint2_com, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(up_Zhewan1, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(down_Zhewan1, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(up_Zhongwan6, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(down_Zhongwan6, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(up_10SpriZhe, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(down_10SpriZhe, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(up_10SpriZho, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(down_10SpriZho, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(up_20SpriZhe, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(down_20SpriZhe, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(up_20SpriZho, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(down_20SpriZho, x, 0.01, full_mapping_ext, geneUniverse_ext))


sapply(ls()[grepl('^((up)|(down))_', ls(), perl = T)], function(x) x %>% get() %>% 
         write.table(file.path('revision_round1_supplementaries' ,'DE_GO_tables', paste(x, 'tsv', sep='.')), sep = '\t', row.names = F, col.names = T))

sapply(ls()[grepl('((BP)|(CC)|MF)_((up)|(down))_', ls(), perl = T)], function(x) x %>% get() %>% 
         write.table(file.path('revision_round1_supplementaries' ,'DE_GO_tables', paste(x, 'tsv', sep='.')), sep = '\t', row.names = F, col.names = T))




bars_plot_n_save <- function(up, down, cutoff, path, w=48, h=20) {
 tiff(filename = path,
      width = w, height = h, units = "cm", pointsize = 12,
      compression = "lzw",
      bg = "white", res = 520, family = "",
      type = "cairo") 

  print(GO_barplotter(up, down, cutoff))
  
  dev.off()
    
}

bars_plot_n_save(BP_up_Sprint2, BP_down_Sprint2, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Sprint2', 'BP.tiff'))
bars_plot_n_save(CC_up_Sprint2, CC_down_Sprint2, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Sprint2', 'СС.tiff'))
bars_plot_n_save(MF_up_Sprint2, MF_down_Sprint2, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Sprint2', 'MF.tiff'), w = 72)
bars_plot_n_save(BP_up_Sprint2_com, BP_down_Sprint2_com, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Sprint2_combined', 'BP.tiff'))
bars_plot_n_save(CC_up_Sprint2_com, CC_down_Sprint2_com, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Sprint2_combined', 'CC.tiff'))
bars_plot_n_save(MF_up_Sprint2_com, MF_down_Sprint2_com, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Sprint2_combined', 'MF.tiff'))
bars_plot_n_save(BP_up_Zhewan1, BP_down_Zhewan1, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Zhewan1', 'BP.tiff'))
bars_plot_n_save(CC_up_Zhewan1, CC_down_Zhewan1, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Zhewan1', 'CC.tiff'))
bars_plot_n_save(MF_up_Zhewan1, MF_down_Zhewan1, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Zhewan1', 'MF.tiff'))
bars_plot_n_save(BP_up_Zhongwan6, BP_down_Zhongwan6, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Zhongwan6', 'BP.tiff'))
bars_plot_n_save(CC_up_Zhongwan6, CC_down_Zhongwan6, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Zhongwan6', 'CC.tiff'))
bars_plot_n_save(MF_up_Zhongwan6, MF_down_Zhongwan6, 10, file.path('revision_round1_supplementaries', 'GO_barplots', 'Zhongwan6', 'MF.tiff'))
bars_plot_n_save(BP_up_10SpriZhe, BP_down_10SpriZhe, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '10DAP', 'BP_Zhewan.tiff'))
bars_plot_n_save(CC_up_10SpriZhe, CC_down_10SpriZhe, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '10DAP', 'CC_Zhewan.tiff'))
bars_plot_n_save(MF_up_10SpriZhe, MF_down_10SpriZhe, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '10DAP', 'MF_Zhewan.tiff'))
bars_plot_n_save(BP_up_10SpriZho, BP_down_10SpriZho, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '10DAP', 'BP_Zhongwan.tiff'))
bars_plot_n_save(CC_up_10SpriZho, CC_down_10SpriZho, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '10DAP', 'CC_Zhongwan.tiff'))
bars_plot_n_save(MF_up_10SpriZho, MF_down_10SpriZho, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '10DAP', 'MF_Zhongwan.tiff'))
bars_plot_n_save(BP_up_20SpriZhe, BP_down_20SpriZhe, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '20DAP', 'BP_Zhewan.tiff'))
bars_plot_n_save(CC_up_20SpriZhe, CC_down_20SpriZhe, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '20DAP', 'CC_Zhewan.tiff'))
bars_plot_n_save(MF_up_20SpriZhe, MF_down_20SpriZhe, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '20DAP', 'MF_Zhewan.tiff'))
bars_plot_n_save(BP_up_20SpriZho, BP_down_20SpriZho, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '20DAP', 'BP_Zhongwan.tiff'))
bars_plot_n_save(CC_up_20SpriZho, CC_down_20SpriZho, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '20DAP', 'CC_Zhongwan.tiff'))
bars_plot_n_save(MF_up_20SpriZho, MF_down_20SpriZho, 10, file.path('revision_round1_supplementaries', 'GO_barplots', '20DAP', 'MF_Zhongwan.tiff'))

bars_plot_n_save(NA, BP_SdZhedZhod, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'SdZhedZhod', 'BP_SdZhedZhod.tiff'))
bars_plot_n_save(NA, CC_SdZhedZhod, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'SdZhedZhod', 'CC_SdZhedZhod.tiff'))
bars_plot_n_save(NA, MF_SdZhedZhod, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'SdZhedZhod', 'MF_SdZhedZhod.tiff'))

bars_plot_n_save(NA, BP_ZhedZhod, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod', 'BP_ZhedZhod.tiff'))
bars_plot_n_save(NA, CC_ZhedZhod, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod', 'CC_ZhedZhod.tiff'))
bars_plot_n_save(NA, MF_ZhedZhod, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod', 'MF_ZhedZhod.tiff'))

bars_plot_n_save(BP_ZheuZhou, NA, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou', 'A.tiff'))
bars_plot_n_save(CC_ZheuZhou, NA, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou', 'B.tiff'), h = 13)
bars_plot_n_save(MF_ZheuZhou, NA, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou', 'C.tiff'), w = 80)

bars_plot_n_save(NA, BP_ZhedZhod20, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod20', 'A.tiff'))
bars_plot_n_save(NA, CC_ZhedZhod20, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod20', 'B.tiff'))
bars_plot_n_save(NA, MF_ZhedZhod20, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZhedZhod20', 'C.tiff'))

bars_plot_n_save(BP_ZheuZhou_20, NA, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou_20', 'A.tiff'))
bars_plot_n_save(CC_ZheuZhou_20, NA, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou_20', 'B.tiff'))
bars_plot_n_save(MF_ZheuZhou_20, NA, cutoff = 20, path=file.path('revision_round1_supplementaries', 'common_transcripts_GO', 'ZheuZhou_20', 'C.tiff'))


```

### Gene-wise lollipop plots

```{r lollipops, echo=TRUE, eval=TRUE, warning=FALSE, error=FALSE, message=FALSE}

retrieve_betas <- function(sleuthset, checklist=list(ABI3, LEC1, LEC2, FUS3, CMT, DRM, RDM1)) {
  out <- c()
  outstat <- c()
  for (variable in checklist) {
    for (gene in variable) {
  ans <- ifelse(gene %in% sleuthset$target_id, sleuthset %>% filter(target_id == gene) %>% dplyr::select(b) %>% unlist() %>% unname(), NA)
      # out <- c(out, sleuthset %>% filter(target_id %in% variable) %>% dplyr::select(b) %>% unlist() %>% unname())
  p <- ifelse(gene %in% sleuthset$target_id, sleuthset %>% filter(target_id == gene) %>% dplyr::select(qval) %>% unlist() %>% unname(), NA)
  out <- c(out, ans)
  outstat <- c(outstat, p)
    }
  }
  return(list(out %>% unlist(), outstat %>% unlist()))
}

retrieve_betas(joint_results_orig)


lolli_data <- data.frame('gene' = rep(c('ABI3', 'LEC1', 'LEC2', 'FUS3', 'CMT', 'CMT [2]', 'DRM', 'DRM [2]', 'DRM [3]', 'DRM [4]', 'RDM1'), 8), 
                         'condition' = c(rep('Sprint2_native', 11), rep('Sprint2', 11), rep('Zhewan1', 11), rep('Zhongwan6', 11),
                                         rep('10SpriZhe', 11), rep('10SpriZho', 11), rep('20SpriZhe', 11), rep('20SpriZho', 11)),
                         'beta' = c(retrieve_betas(joint_results_orig)[1], retrieve_betas(joint_results_com_orig)[1], retrieve_betas(joint_results_Zhewan1_orig)[1], retrieve_betas(joint_results_Zhongwan6_orig)[1], retrieve_betas(joint_results_10SpriZhe_orig)[1], retrieve_betas(joint_results_10SpriZho_orig)[1], retrieve_betas(joint_results_20SpriZhe_orig)[1], retrieve_betas(joint_results_20SpriZho_orig)[1]) %>% unlist(),
                         'qval' = c(retrieve_betas(joint_results_orig)[2], retrieve_betas(joint_results_com_orig)[2], retrieve_betas(joint_results_Zhewan1_orig)[2], retrieve_betas(joint_results_Zhongwan6_orig)[2], retrieve_betas(joint_results_10SpriZhe_orig)[2], retrieve_betas(joint_results_10SpriZho_orig)[2], retrieve_betas(joint_results_20SpriZhe_orig)[2], retrieve_betas(joint_results_20SpriZho_orig)[2]) %>% unlist()) %>% `colnames<-`(c('gene', 'condition', 'beta', 'qval')) %>% mutate('significance' = ifelse(abs(beta) > 0.5 & qval < 0.001, 'yes', 'no'))

tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'TFs', 'Sprint-2.tiff'),
     width = 16, height = 14, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

ggplot(lolli_data %>% filter(condition == 'Sprint2_native', !is.na(significance)), aes(x = gene, y = beta, fill = qval, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = 0, yend = beta)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_gradientn(colors = colors) +
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.x.bottom = element_text(size = 10, color = 'black', family = 'arial',
                                     face = 'italic', angle = 45, vjust = 0.95, hjust = 0.95),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()

lolli_10days <- lolli_data %>% filter(condition == '10SpriZhe' | condition == '10SpriZho') %>% mutate(beta1 = c(beta[12:22], beta[1:11]))


tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'TFs', '10DAP.tiff'),
     width = 18, height = 14, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

ggplot(lolli_10days %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene %>% unique() %>% rev())), aes(x = gene, y = beta, fill = condition, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = beta, yend = beta1)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_manual(values = c('10SpriZhe' = '#068013', '10SpriZho' = '#9c1718'), name = 'Cultivar',
                      guide = guide_legend(override.aes = list(shape=21)), labels = c('Zhewan-1', 'Zhongwan-6')) +
    coord_flip() + 
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.x = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial', face = 'italic'),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')


dev.off()

lolli_20days <- lolli_data %>% filter(condition == '20SpriZhe' | condition == '20SpriZho') %>% 
  mutate(beta1 = c(beta[12:22], beta[1:11])) %>% filter_at(vars(c(significance)), all_vars(!is.na(.)))


tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', '20DAP.tiff'),
     width = 18, height = 14, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")
    
ggplot(lolli_20days %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene %>% unique() %>% rev())), aes(x = gene, y = beta, fill = condition, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = beta, yend = beta1)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_manual(values = c('20SpriZhe' = '#068013', '20SpriZho' = '#9c1718'), name = 'Cultivar',
                      guide = guide_legend(override.aes = list(shape=21)), labels = c('Zhewan-1', 'Zhongwan-6')) +
    coord_flip() + 
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.x = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial', face = 'italic'),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()

```


### Transcription factor tracking

More than 300 entries related to transcription regulation (including chromatin consituent and mRNA splicing) seem to have been discordantly annotated by MapMan and BLAST2GO/eggNOG. To further reinforce TF trafficking we download annotations for these contigs to check their real functions manually.

```{r TFs_DE, echo=TRUE, eval=TRUE, warning=FALSE, error=FALSE, message=FALSE}

descrepant_TFs <- annot_full_ext %>% filter(grepl('regulation of transcription', MapMan_term), !grepl(paste0('(', paste(c(TF_GO, chromatin, splicing), collapse = ')|('), ')', sep = ''), GO_BLAST),
                      !grepl('GO:0005634', GO_BLAST)) %>% dplyr::select(transcript_id, prot_id, GO_BLAST, MapMan_term, BLASTP_names, BLASTX_names)

descrepant_TFs$GO_BLAST <- unlist(descrepant_TFs$GO_BLAST)

TF_matrix <- annot_full_ext %>% filter(grepl(paste0('(', paste(c(TF_GO, chromatin, splicing, 'GO:0005634'), collapse = ')|('), ')', sep = ''), GO_BLAST)) %>% dplyr::select(transcript_id)

# a conciser and more accurateversion depicting only MapMan-annotated transcription factors
TF_matrix <- annot_full_ext %>% filter(grepl('regulation of transcription', MapMan_term)) %>% dplyr::select(transcript_id)

TF_matrix <-   cbind(TF_matrix, joint_results_com[match(TF_matrix$transcript_id, joint_results_com$target_id),]$b,
                   joint_results_Zhewan1[match(TF_matrix$transcript_id,joint_results_Zhewan1$target_id),]$b,
                   joint_results_Zhongwan6[match(TF_matrix$transcript_id,joint_results_Zhongwan6$target_id),]$b,
                   joint_results_10SpriZhe[match(TF_matrix$transcript_id,joint_results_10SpriZhe$target_id),]$b,
                   joint_results_10SpriZho[match(TF_matrix$transcript_id, joint_results_10SpriZho$target_id),]$b,
                   joint_results_20SpriZhe[match(TF_matrix$transcript_id, joint_results_20SpriZhe$target_id),]$b,
                   joint_results_20SpriZho[match(TF_matrix$transcript_id, joint_results_20SpriZho$target_id),]$b) %>% column_to_rownames('transcript_id')

colnames(TF_matrix) <- c('Sprint-2,10 vs Sprint-2,20', 'Zhewan-1,10 vs Zhewan-1,25', 'Zhongwan-6,10 vs Zhongwan-6,25', 'Sprint-2,10 vs Zhewan-1,10',
                         'Sprint-2,10 vs Zhongwan-6,10', 'Sprint-2,20 vs Zhewan-1,25', 'Sprint-2,20 vs Zhongwan-6,25')

TF_matrix[is.na(TF_matrix)] <- 0
TF_matrix %<>% filter_all(any_vars(. != 0))


colors <- colorRampPalette( rev(brewer.pal(9, 'RdYlGn')) )(255)

tiff(filename = file.path('revision_round1_supplementaries', 'heatmaps', 'TFs.tiff'),
     width = 26, height = 24, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

pheatmap(TF_matrix, show_rownames = F, color = colors, cluster_rows = F, cluster_cols = T, fontsize = 12)

dev.off()


tiff(filename = file.path('revision_round1_supplementaries', 'heatmaps', 'TFs_reordered.tiff'),
     width = 28, height = 24, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

TF_matrix %<>% arrange(desc(`Sprint-2,10 vs Sprint-2,20`))

pheatmap(TF_matrix, show_rownames = F, color = colors, cluster_rows = F, cluster_cols = T, fontsize = 12)

dev.off()


write.table(descrepant_TFs, file.path('auxiliary_materials', 'descrepant_TFs.tsv'), row.names = F, col.names = T, sep = '\t')

hormone_matrix <- annot_full_ext %>% filter(grepl(paste0('(', paste(c(TF_GO, hormones), collapse = ')|('), ')', sep = ''), GO_BLAST)) %>% dplyr::select(transcript_id)

# just again, a MapMan-restricted subselection

hormone_matrix <- annot_full_ext %>% filter(grepl('hormone', MapMan_term)) %>% dplyr::select(transcript_id)

hormone_matrix <-   cbind(hormone_matrix, joint_results_com[match(hormone_matrix$transcript_id, joint_results_com$target_id),]$b,
                  joint_results_Zhewan1[match(hormone_matrix$transcript_id,joint_results_Zhewan1$target_id),]$b,
                  joint_results_Zhongwan6[match(hormone_matrix$transcript_id,joint_results_Zhongwan6$target_id),]$b,
                   joint_results_10SpriZhe[match(hormone_matrix$transcript_id,joint_results_10SpriZhe$target_id),]$b,
                   joint_results_10SpriZho[match(hormone_matrix$transcript_id, joint_results_10SpriZho$target_id),]$b,
                   joint_results_20SpriZhe[match(hormone_matrix$transcript_id, joint_results_20SpriZhe$target_id),]$b,
                   joint_results_20SpriZho[match(hormone_matrix$transcript_id, joint_results_20SpriZho$target_id),]$b) %>% column_to_rownames('transcript_id')

colnames(hormone_matrix) <- c('Sprint-2,10 vs Sprint-2,20', 'Zhewan-1,10 vs Zhewan-1,25', 'Zhongwan-6,10 vs Zhongwan-6,25', 'Sprint-2,10 vs Zhewan-1, 10', 'Sprint-2,10 vs Zhongwan-6,10', 'Sprint-2,20 vs Zhewan-1,25', 'Sprint-2,20 vs Zhongwan-6,25')

hormone_matrix[is.na(hormone_matrix)] <- 0
hormone_matrix %<>% filter_all(any_vars(. != 0))
hormone_matrix %<>% arrange(desc(`Sprint-2,10 vs Sprint-2,20`))


draw_colnames_45 <- function (coln, gaps, ...) {
    coord <-  pheatmap:::find_coordinates(length(coln), gaps)
    x <-  coord$coord - 0.5 * coord$size
    res <-  textGrob(coln, x = x, y = unit(1, "npc") - unit(3,"bigpts"), vjust = 3, hjust = 0.1, rot = -45, gp = gpar(...))
    return(res)}
assignInNamespace(x="draw_colnames", value="draw_colnames_45", ns=asNamespace("pheatmap"))


tiff(filename = file.path('revision_round1_supplementaries', 'heatmaps', 'hormones.tiff'),
     width = 28, height = 24, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

pheatmap(hormone_matrix, show_rownames = F, color = colors, cluster_rows = F, cluster_cols = T, fontsize = 12)


dev.off()

tiff(filename = file.path('revision_round1_supplementaries', 'heatmaps', 'hormones_reordered.tiff'),
     width = 28, height = 24, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

hormone_matrix %<>% arrange(desc(`Sprint-2,10 vs Sprint-2,20`))

pheatmap(hormone_matrix, show_rownames = F, color = colors, cluster_rows = F, cluster_cols = T, fontsize = 12)


dev.off()



```

#### TF-based hormone-related DEG tracking

```{r heatmap_clusters_TF, , echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}
Sprint2_down_TF_5 <- annot_full_ext %>% filter(transcript_id %in% down_Sprint2_com[down_Sprint2_com$b < -5,]$target_id, grepl('regulation of transcription', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)

Sprint2_up_TF_5 <- annot_full_ext %>% filter(transcript_id %in% up_Sprint2_com[up_Sprint2_com$b > 5,]$target_id, grepl('regulation of transcription', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
Sprint2_down_TF_5 %>% View()
```

  

#### Heatmap-based hormone-related DEG tracking

```{r heatmap_clusters_horm, , echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}
Sprint2_down_hormone_5 <- annot_full_ext %>% filter(transcript_id %in% down_Sprint2_com[down_Sprint2_com$b < -5,]$target_id, grepl('hormone', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)

Sprint2_up_hormone_5 <- annot_full_ext %>% filter(transcript_id %in% up_Sprint2_com[up_Sprint2_com$b > 5,]$target_id, grepl('hormone', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
Sprint2_up_hormone_5 %>% View()
```

#### BLAST-based hormone-related DEG tracking

```{r BLAST_candidates_gibb, , echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

GA_refs <- read.table(file.path('GA_refs', 'GA_annotations_060220_metabolism_only.ssv'), sep = ';', header = T, stringsAsFactors = F, dec = '.', quote = '') %>% set_colnames(c('Contig', 'Accession', 'Description')) %>% filter(!duplicated(Contig)) %>%  mutate('Name' = sapply(Description, function(x) gsub('[XN]P_[.0-9]+ ', '', x) %>% gsub('\\[[A-Za-z ]+\\]', '', .)))

```

The list above contains homologs identified via BLAST against a database comprising of RefSeq references. This approach, however, omits paralogs and/or spliceforms since only the top 1 matches were selected by the primary BLAST statistics (see *ipynb* notebook for respective *sh* code chunks) for each enzyme. So, in the hindsight several other transcripts were added. The commented lines contain code which was used for finding these missing homologs.

```{r BLAST_candidates_gibb_2, , echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

# annot_full_ext %>% filter(grepl('ent-kaurenoic acid oxidase', BLASTP_names) | grepl('ent-kaurenoic acid oxidase', BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('hormone metabolism.gibberelin.synthesis-degradation.ent-kaurenoic acid hydroxylase/oxygenase', MapMan_term))
#annot_full_ext %>% filter(transcript_id == 'TRINITY_DN6183_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
#annot_full_ext %>% filter(grepl('copalyl diphosphate synthase', BLASTP_names) | grepl('copalyl diphosphate synthase', BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN3182_c1_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('hormone metabolism.gibberelin.synthesis-degradation.GA20 oxidase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
#annot_full_ext %>% filter(transcript_id == 'TRINITY_DN12474_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('hormone metabolism.gibberelin.synthesis-degradation.ent-kaurene synthase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN1657_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('cytochrome P450 714A1', BLASTP_names) | grepl('cytochrome P450 714A1', BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full %>% filter(transcript_id == 'TRINITY_DN28727_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('hormone metabolism.gibberelin.synthesis-degradation.GA3 oxidase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full %>% filter(transcript_id == 'TRINITY_DN6286_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('hormone metabolism.gibberelin.synthesis-degradation.GA2 oxidase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)


GA_refs %<>% rbind(data.frame('Contig' = c('TRINITY_DN5289_c0_g2_i1', 'TRINITY_DN13727_c0_g1_i3', 'TRINITY_DN6183_c0_g1_i2',
                                           'TRINITY_DN8852_c0_g4_i1', 'TRINITY_DN8852_c0_g3_i2', 'TRINITY_DN10882_c0_g1_i1',
                                           'TRINITY_DN52652_c0_g1_i1', 'TRINITY_DN18761_c0_g1_i1',
                                           'comp48083_c0_seq1', 'TRINITY_DN6518_c0_g1_i1', 'TRINITY_DN5711_c0_g1_i2',
                                           'TRINITY_DN22573_c0_g2_i1', 'TRINITY_DN6006_c0_g1_i2'),
                              'Accession' = c('NP_001315754.1', 'NP_001234008.2', 'NP_001234008.2', 'XP_003528896.1', 'XP_003528896.1',
                                              'XP_003528896.1', 'XP_003528896.1', 'XP_003528896.1', 'XP_003528896.1', 
                                              'XP_015645488.1', 'XP_022977872.1', 'XP_022977872.1', 'XP_022977872.1'),
                              'Description' = c('NP_001315754.1 ent-kaurenoic acid oxidase 1-like [Malus domestica]', 'NP_001234008.2 copalyl diphosphate synthase [Solanum lycopersicum]', 'NP_001234008.2 copalyl diphosphate synthase [Solanum lycopersicum]', 'XP_003528896.1 gibberellin 20 oxidase 2 [Glycine max]', 'XP_003528896.1 gibberellin 20 oxidase 2 [Glycine max]', 'XP_003528896.1 gibberellin 20 oxidase 2 [Glycine max]', 'XP_003528896.1 gibberellin 20 oxidase 2 [Glycine max]', 'XP_003528896.1 gibberellin 20 oxidase 2 [Glycine max]', 'XP_003528896.1 gibberellin 20 oxidase 2 [Glycine max]', 'XP_015645488.1 cytochrome P450 714B1 [Oryza sativa Japonica Group]', 'XP_022977872.1 gibberellin 2-beta-dioxygenase [Cucurbita maxima]', 'XP_022977872.1 gibberellin 2-beta-dioxygenase [Cucurbita maxima]', 'XP_022977872.1 gibberellin 2-beta-dioxygenase [Cucurbita maxima]'),
                              'Name' = c('ent-kaurenoic acid oxidase 1-like[2]', 'copalyl diphosphate synthase[2]', 'copalyl diphosphate synthase[3]', 'gibberellin 20 oxidase 2[2]', 'gibberellin 20 oxidase 2[3]', 'gibberellin 20 oxidase 2[4]', 'gibberellin 20 oxidase 2[5]', 'gibberellin 20 oxidase 2[6]', 'gibberellin 20 oxidase 2[7]', 'cytochrome P450 714B1[2]', 'gibberellin 2-beta-dioxygenase[2]', 'gibberellin 2-beta-dioxygenase[3]', 'gibberellin 2-beta-dioxygenase[4]'))) %>% mutate(Name = Name %>% gsub('copalyl diphosphate synthase( )?', 'CPS', ., perl = T) %>% 
gsub('ent-kaur-16-ene synthase, chloroplastic isoform X1( )?', 'KS', ., perl = T) %>%
  gsub('GA requiring 3( )?', 'KO', ., perl = T) %>%
  gsub('ent-kaurenoic acid oxidase 1-like( )?', 'KAO', ., perl = T) %>%
  gsub('cytochrome P450 714B1( )?', 'GA13ox', ., perl = T) %>%
  gsub('gibberellin 20 oxidase 2( )?', 'GA20ox', ., perl = T) %>%
  gsub('PREDICTED: gibberellin 3-beta-dioxygenase 1-like( )?', 'GA3ox', ., perl = T) %>%
  gsub('PREDICTED: gibberellin 3-beta-dioxygenase 1-like( )?', 'GA3ox', ., perl = T) %>%
  gsub('gibberellin 2-beta-dioxygenase( )?', 'GA2ox', ., perl = T) %>% 
  gsub('\\[', ' \\[', ., perl = T)) 

GA_lolli <- data.frame('gene' = rep(GA_refs$Name, 8), 
                       'condition' = c(rep('Sprint2_native', nrow(GA_refs)), rep('Sprint2', nrow(GA_refs)), rep('Zhewan1', nrow(GA_refs)), rep('Zhongwan6', nrow(GA_refs)),
                                         rep('10SpriZhe', nrow(GA_refs)), rep('10SpriZho', nrow(GA_refs)), rep('20SpriZhe', nrow(GA_refs)), rep('20SpriZho', nrow(GA_refs))),
                       'beta' = sapply(list(joint_results_orig, joint_results_com_orig, joint_results_Zhewan1_orig, joint_results_Zhongwan6_orig, joint_results_10SpriZhe_orig, joint_results_10SpriZho_orig, joint_results_20SpriZhe_orig, joint_results_20SpriZho_orig), function(x) retrieve_betas(x, GA_refs$Contig) %>% dplyr::first()) %>% as.vector(),
                         'qval' = sapply(list(joint_results_orig, joint_results_com_orig, joint_results_Zhewan1_orig, joint_results_Zhongwan6_orig, joint_results_10SpriZhe_orig, joint_results_10SpriZho_orig, joint_results_20SpriZhe_orig, joint_results_20SpriZho_orig), function(x) retrieve_betas(x, GA_refs$Contig) %>% dplyr::nth(2)) %>%
                         as.vector()) %>% 
  mutate('significance' = ifelse(abs(beta) > 0.5 & qval < 0.001, 'yes', 'no'))


tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'hormones', 'GA_Sprint.tff'),
     width = 16, height = 14, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

ggplot(GA_lolli %>% filter(condition == 'Sprint2_native', !is.na(significance)) %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene)), aes(x = gene, y = beta, fill = qval, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = 0, yend = beta)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_gradientn(colors = colors) +
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.x.bottom = element_text(size = 10, color = 'black', family = 'arial',
                                     face = 'italic', angle = 45, hjust = 0.95, vjust = 0.95),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()

GA_lolli_10days <- GA_lolli %>% filter(condition == '10SpriZhe' | condition == '10SpriZho') %>% mutate(beta1 = c(beta[22:42], beta[1:21])) %>% filter(!is.na(significance))


tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'hormones', 'GA_10days.tff'),
     width = 16, height = 14, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")


ggplot(GA_lolli_10days %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene %>% unique() %>% rev())), aes(x = gene, y = beta, fill = condition, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = beta, yend = beta1)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_manual(values = c('10SpriZhe' = '#068013', '10SpriZho' = '#9c1718'), name = 'Cultivar',
                      guide = guide_legend(override.aes = list(shape=21)), labels = c('Zhewan-1', 'Zhongwan-6')) +
    coord_flip() + 
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.x = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial', face = 'italic'),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()


GA_lolli_20days <- GA_lolli %>% filter(condition == '20SpriZhe' | condition == '20SpriZho') %>% mutate(beta1 = c(beta[22:42], beta[1:21])) %>% filter(!is.na(significance))


tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'hormones', 'GA_20days.tff'),
     width = 16, height = 14, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")


ggplot(GA_lolli_20days %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene %>% unique() %>% rev())), aes(x = gene, y = beta, fill = condition, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = beta, yend = beta1)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_manual(values = c('20SpriZhe' = '#068013', '20SpriZho' = '#9c1718'), name = 'Cultivar',
                      guide = guide_legend(override.aes = list(shape=21)), labels = c('Zhewan-1', 'Zhongwan-6')) +
    coord_flip() + 
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.x = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial', face = 'italic'),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()



```

Then we apply the same approach to the ABA biosynthesis-related transcripts:

```{r BLAST_candidates_aba, , echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}
ABA_refs <- read.table(file.path('ABA_refs', 'ABA_annotations_060220_metabolism_only.ssv'), sep = ';', header = T, stringsAsFactors = F, dec = '.', quote = '') %>% set_colnames(c('Contig', 'Accession', 'Description')) %>% filter(!duplicated(Contig)) %>%  mutate('Name' = sapply(Description, function(x) gsub('[XN]P_[.0-9]+ ', '', x) %>% gsub('\\[[A-Za-z ]+\\]', '', .)))
```

As in the case of GA biosynthesis, the commented lines illustrate the logic for isoform/paralog dissection and inclusion

```{r BLAST_candidates_aba_2, , echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN5423_c0_g1_i3') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('1-deoxy-D-xylulose-5-phosphate synthase', BLASTP_names) | grepl('1-deoxy-D-xylulose-5-phosphate synthase', BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN2259_c0_g2_i3') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('phytoene synthase 2', BLASTP_names) | grepl('1-deoxy-D-xylulose-5-phosphate synthase', BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN8432_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('secondary metabolism.isoprenoids.carotenoids.zeta-carotene desaturase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN13325_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('secondary metabolism.isoprenoids.carotenoids.phytoene dehydrogenase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN1144_c0_g2_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('secondary metabolism.isoprenoids.carotenoids.violaxanthin de-epoxidase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN525_c0_g1_i3') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('secondary metabolism.isoprenoids.carotenoids.carotenoid isomerase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN5462_c0_g1_i3') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('molybdenum cofactor sulfurase', BLASTP_names) | grepl('molybdenum cofactor sulfurase', BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN6988_c0_g1_i4') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('aldehyde oxidase 3', BLASTP_names) | grepl('aldehyde oxidase 3', BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN12636_c0_g1_i2') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl("abscisic acid 8'-hydroxylase", BLASTP_names) | grepl("abscisic acid 8'-hydroxylase", BLASTX_names)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('(ABA2)', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN2857_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('secondary metabolism.isoprenoids.carotenoids.lycopene beta cyclase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN9199_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('secondary metabolism.isoprenoids.non-mevalonate pathway.geranylgeranyl pyrophosphate synthase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(transcript_id == 'TRINITY_DN4936_c0_g1_i1') %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)
# annot_full_ext %>% filter(grepl('hormone metabolism.abscisic acid.synthesis-degradation.synthesis.zeaxanthin epoxidase', MapMan_term)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names, MapMan_term)


ABA_refs %<>% rbind(data.frame('Contig' = c('TRINITY_DN652_c0_g1_i4', 'TRINITY_DN5423_c0_g2_i1', 'TRINITY_DN7540_c0_g3_i1', 'TRINITY_DN18026_c0_g1_i1', 'TRINITY_DN11514_c0_g1_i1', 'TRINITY_DN782_c0_g1_i1', 'TRINITY_DN6270_c0_g1_i1', 'TRINITY_DN33839_c0_g1_i1', 'TRINITY_DN19208_c0_g1_i1', 'TRINITY_DN55409_c0_g1_i1', 'TRINITY_DN41836_c0_g1_i1','TRINITY_DN24587_c0_g1_i1', 'TRINITY_DN30523_c0_g1_i1', 'TRINITY_DN6988_c0_g1_i6', 'TRINITY_DN8814_c0_g1_i7', 'TRINITY_DN24427_c0_g1_i1', 'TRINITY_DN18407_c0_g2_i1', 'TRINITY_DN52078_c0_g1_i1', 'TRINITY_DN12553_c1_g1_i1', 'TRINITY_DN12553_c0_g1_i2', 'TRINITY_DN12553_c0_g1_i3', 'TRINITY_DN16303_c0_g1_i2', 'TRINITY_DN16303_c0_g1_i3', 'TRINITY_DN19193_c0_g1_i1', 'TRINITY_DN4936_c0_g1_i5', 'TRINITY_DN5762_c0_g1_i2'),
                               'Accession' = c('NP_001119420.1', rep('NP_001234672.1', 3), 'NP_001318580.1',
                                               rep('NP_001321301.1', 2), rep('NP_001322902.1', 6), 'NP_001324269.1',
                                               rep('NP_001324571.1', 7), rep('NP_195399.1', 3), 'NP_201504.2', 'NP_564889.1'),
                               'Description' = c('NP_001119420.1 beta-carotene hydroxylase 2 [Arabidopsis thaliana]', rep('NP_001234672.1 1-D-deoxyxylulose 5-phosphate synthase [Solanum lycopersicum]', 3), 'NP_001318580.1 PHYTOENE SYNTHASE [Arabidopsis thaliana]', rep('NP_001321301.1 non-photochemical quenching 1 [Arabidopsis thaliana]', 2), rep('NP_001322902.1 molybdenum cofactor sulfurase (LOS5) (ABA3)', 6), 'NP_001324269.1 abscisic aldehyde oxidase 3 [Arabidopsis thaliana]', rep('NP_001324571.1 cytochrome P450, family 707, subfamily A, polypeptide 2 [Arabidopsis thaliana]', 7), rep('NP_195399.1 geranylgeranyl pyrophosphate synthase 1 [Arabidopsis thaliana]', 3), 'NP_201504.2 zeaxanthin epoxidase (ZEP) (ABA1) [Arabidopsis thaliana]', 'NP_564889.1 abscisic acid (aba)-deficient 4 [Arabidopsis thaliana]'),
                               'Name' = c('beta-carotene hydroxylase 2[2]', sapply(2:4, function(x) paste('1-D-deoxyxylulose 5-phosphate synthase', '[', x, ']', sep = '')), 'PHYTOENE SYNTHASE[2]', 'non-photochemical quenching 1[2]', 'non-photochemical quenching 1[3]',  sapply(2:7, function(x) paste('molybdenum cofactor sulfurase (LOS5) (ABA3)', '[', x, ']', sep = '')), 'abscisic aldehyde oxidase 3[2]', sapply(2:8, function(x) paste('cytochrome P450, family 707, subfamily A, polypeptide 2', '[', x, ']', sep ='')), sapply(2:4, function(x) paste('geranylgeranyl pyrophosphate synthase 1', '[', x, ']', sep = '')), ' zeaxanthin epoxidase (ZEP) (ABA1)', 'abscisic acid (aba)-deficient 4 [2]'))) %>% mutate(Name = Name %>% gsub('zeaxanthin epoxidase \\(ZEP\\) \\(ABA1\\)( )?', 'ABA1', ., perl = T) %>%
                                                gsub('non-photochemical quenching 1( )?', 'VDE1', ., perl = T) %>% 
                                                gsub('abscisic acid \\(aba\\)-deficient 4( )?', 'ABA4', ., perl = T) %>% 
                                                gsub('NAD\\(P\\)-binding Rossmann-fold superfamily protein( )?', 'ABA2', ., perl = T) %>% 
                                                gsub('aldehyde oxidase 1( )?', 'AAO1', ., perl = T) %>% 
                                                gsub('abscisic aldehyde oxidase 3( )?', 'AAO3', ., perl = T) %>% 
                                                gsub('abscisic aldehyde oxidase 4( )?', 'AAO4', ., perl = T) %>% 
                                                gsub('beta-hydroxylase 1( )?', 'BCH1', ., perl = T) %>% 
                                                gsub('beta-carotene hydroxylase 2( )?', 'SCH2', ., perl = T) %>% 
                                                gsub('carotenoid isomerase( )?', 'CRTISO', ., perl = T) %>% 
                                                gsub('1-D-deoxyxylulose 5-phosphate synthase( )?', 'DXS', ., perl = T) %>% 
                                                gsub('cytochrome P450, family 707, subfamily A, polypeptide 1', 'CYP707A1', ., perl = T) %>% 
                                                gsub('cytochrome P450, family 707, subfamily A, polypeptide 3( )?' , 'CYP707A3', ., perl = T) %>% 
                                                gsub('cytochrome P450, family 707, subfamily A, polypeptide 2( )?', 'CYP707A2', ., perl = T) %>% 
                                                gsub('lycopene cyclase( )?', 'LCYB', ., perl = T) %>% 
                                                gsub('phytoene desaturase 3( )?', 'PDS', ., perl = T) %>% 
                                                gsub('PHYTOENE SYNTHASE( )?', 'PSY', ., perl = T) %>% 
                                                gsub('geranylgeranyl pyrophosphate synthase 1( )?', 'GGPS', ., perl = T) %>% 
                                                gsub('zeta-carotene desaturase( )?', 'ZDS1', ., perl = T) %>% 
                                                gsub('molybdenum cofactor sulfurase \\(LOS5\\) \\(ABA3\\)( )?', 'ABA3', ., perl = T) %>% gsub('\\[', ' \\[', .))

ABA_lolli <- data.frame('gene' = rep(ABA_refs$Name, 8), 
                       'condition' = c(rep('Sprint2_native', nrow(ABA_refs)), rep('Sprint2', nrow(ABA_refs)), rep('Zhewan1', nrow(ABA_refs)), rep('Zhongwan6', nrow(ABA_refs)),
                                         rep('10SpriZhe', nrow(ABA_refs)), rep('10SpriZho', nrow(ABA_refs)), rep('20SpriZhe', nrow(ABA_refs)), rep('20SpriZho', nrow(ABA_refs))),
                       'beta' = sapply(list(joint_results_orig, joint_results_com_orig, joint_results_Zhongwan6_orig, joint_results_Zhewan1_orig, joint_results_10SpriZhe_orig, joint_results_10SpriZho_orig, joint_results_20SpriZhe_orig, joint_results_20SpriZho_orig), function(x) retrieve_betas(x, ABA_refs$Contig) %>% dplyr::first()) %>% as.vector(),
                         'qval' = sapply(list(joint_results_orig, joint_results_com_orig, joint_results_Zhongwan6_orig, joint_results_Zhewan1_orig, joint_results_10SpriZhe_orig, joint_results_10SpriZho_orig, joint_results_20SpriZhe_orig, joint_results_20SpriZho_orig), function(x) retrieve_betas(x, ABA_refs$Contig) %>% dplyr::nth(2)) %>% as.vector()) %>% mutate('significance' = ifelse(abs(beta) > 0.5 & qval < 0.001, 'yes', 'no'))


tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'hormones', 'ABA_Sprint2.tff'),
     width = 24, height = 14, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

ggplot(ABA_lolli %>% filter(condition == 'Sprint2_native', !is.na(significance)) %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene %>% unique() %>% rev())), aes(x = gene, y = beta, fill = qval, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = 0, yend = beta)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_gradientn(colors = colors) +
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.x.bottom = element_text(size = 10, color = 'black', family = 'arial',
                                     face = 'italic', angle = 45, vjust = 0.95, hjust = 0.95),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()

ABA_lolli_10days <- ABA_lolli %>% filter(condition == '10SpriZhe' | condition == '10SpriZho') %>% mutate(beta1 = c(beta[43:84], beta[1:42])) %>% filter(!is.na(significance))

tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'hormones', 'ABA_10days.tff'),
     width = 26, height = 18, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

ggplot(ABA_lolli_10days %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene %>% unique() %>% rev())), aes(x = gene, y = beta, fill = condition, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = beta, yend = beta1)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_manual(values = c('10SpriZhe' = '#068013', '10SpriZho' = '#9c1718'), name = 'Cultivar',
                      guide = guide_legend(override.aes = list(shape=21)), labels = c('Zhewan-1', 'Zhongwan-6')) +
    coord_flip() + 
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.x = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial', face = 'italic'),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()

ABA_lolli_20days <- ABA_lolli %>% filter(condition == '20SpriZhe' | condition == '20SpriZho') %>% mutate(beta1 = c(beta[43:84], beta[1:42])) %>% filter(!is.na(significance))

tiff(filename = file.path('revision_round1_supplementaries', 'lollipop_plots', 'hormones', 'ABA_20days.tff'),
     width = 26, height = 18, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

ggplot(ABA_lolli_20days %>% arrange(gene) %>% mutate(gene = factor(gene, levels = gene %>% unique() %>% rev())), aes(x = gene, y = beta, fill = condition, shape = significance))  + 
  geom_segment(aes(x = gene, xend = gene, y = beta, yend = beta1)) + 
  geom_point(size = 6, stroke = 0.5) + 
  scale_shape_manual(values = c('yes' = 21, 'no' = 22), name = 'Statistical significance') + 
  theme_bw() +
  scale_fill_manual(values = c('20SpriZhe' = '#068013', '20SpriZho' = '#9c1718'), name = 'Cultivar',
                      guide = guide_legend(override.aes = list(shape=21)), labels = c('Zhewan-1', 'Zhongwan-6')) +
    coord_flip() + 
    theme(axis.title = element_text(size = 13, color = 'black', family = 'arial'),
          axis.text.x = element_text(size = 10, color = 'black', family = 'arial'),
          axis.text.y = element_text(size = 10, color = 'black', family = 'arial', face = 'italic'),
          legend.text = element_text(color = 'black', family = 'arial'),
          legend.title = element_text(color = 'black', family = 'arial')) +
    labs(y = "lnFC (sleuth's b)", x = '')

dev.off()

```


To fulfill the differential expression block, we export the results of DE and GO over-representation assays to process them into readable spreadsheets:

```{r export_DE_GO, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}
sapply(ls()[grepl('(BP)|(CC)|(MF)', ls())], function(x) x %>% get() %>% write.table(., file.path('GO_BLAST_DE_291119', paste(x, 'tsv', sep='.')), sep = '\t',
                                                                                    col.names = T, row.names = F))

sapply(ls()[grepl('^((up)|(down))', ls(), perl = TRUE)], function(x) x %>% get() %>% write.table(., file.path('DE_291119', paste(x, 'tsv', sep = '.')),
                                                                                                 sep = '\t', col.names = T, row.names = F))

```

## Physiological timeline reconstruction

### Poissonian Clustering
```{r PoClaClu, , echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}
full_matrix <- read.table(file.path('kallisto_outputs_combined_assembly', 'all_trinity_abundance.tsv'),
                          sep = '\t', header = T)

colnames(full_matrix) <- c('target_id', 'Sprint-2,10(1)', 'Sprint-2,10(2)', 'Sprint-2,10(3)', 'Sprint-2,10(4)',
                           'Sprint-2,20(1)', 'Sprint-2,20(2)', 'Sprint-2,20(3)', 'Sprint-2,20(4)',
                           'Zhewan-1,10(1)', 'Zhewan-1,10(2)', 'Zhewan-1,10(3)', 'Zhewan-1,10(4)',
                           'Zhewan-1,25(1)', 'Zhewan-1,25(2)', 'Zhewan-1,25(3)', 'Zhewan-1,25(4)',
                           'Zhongwan-6,10(1)', 'Zhongwan-6,10(2)', 'Zhongwan-6,10(3)', 'Zhongwan-6,10(4)',
                           'Zhongwan-6,25(1)', 'Zhongwan-6,25(2)', 'Zhongwan-6,25(3)', 'Zhongwan-6,25(4)')

poisd <- PoissonDistance(t(full_matrix[,2:25]))
samplePoisDistMatrix <- as.matrix(poisd$dd)
rownames(samplePoisDistMatrix) <- colnames(full_matrix[-1])
colnames(samplePoisDistMatrix) <- samplePoisDistMatrix %>% rownames()
annotation_pois <- data.frame('cultivar' = as.factor(c(rep('Sprint-2', 8), rep('Zhewan-1', 8), rep('Zhongwan-6', 8))),
                              'DAP' = as.factor(c(rep(10, 4), rep(20, 4), rep(10, 4), rep(25, 4), rep(10, 4), rep(25, 4))))
rownames(annotation_pois) <- colnames(full_matrix[2:25])

annot_colors <- list('cultivar' =  c('#068013', '#f1cb20', '#b30a07') %>% `names<-`(c('Sprint-2', 'Zhewan-1', 'Zhongwan-6')),
                     'DAP' = c('#068013', '#f1cb20', '#b30a07') %>% `names<-`(c('10', '20', '25')))

pheatmap(samplePoisDistMatrix,  clustering_distance_rows=poisd$dd,clustering_distance_cols=poisd$dd, col=colors,
         show_rownames = F, show_colnames = F, annotation_col = annotation_pois, annotation_row = annotation_pois,
         annotation_colors = annot_colors)

```

### PCA

```{r PCA, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

for_pca_peas = t(full_matrix[,2:25])
rownames(for_pca_peas) <- colnames(full_matrix)[2:25]
colnames(for_pca_peas) <- full_matrix$target_id
pca_peas <- prcomp(for_pca_peas[,-c(which(apply(for_pca_peas, 2, var) == 0))], center = T, scale. = T)
pca_peas
pca_data <- data.frame('Sample' = c('Sprint-2,10(1)', 'Sprint-2,10(2)', 'Sprint-2,10(3)', 'Sprint-2,10(4)',
                                    'Sprint-2,30(1)', 'Sprint-2,30(2)', 'Sprint-2,30(3)', 'Sprint-2,30(4)',
                                    'Zhewan-1,10(1)', 'Zhewan-1,10(2)', 'Zhewan-1,10(3)', 'Zhewan-1,10(4)',
                                    'Zhewan-1,25(1)', 'Zhewan-1,25(2)', 'Zhewan-1,25(3)', 'Zhewan-1,25(4)',
                                    'Zhongwan-6,10(1)', 'Zhongwan-6,10(1)', 'Zhongwan-6,10(3)', 'Zhongwan-6,10(4)',
                                    'Zhongwan-6,25(1)', 'Zhongwan-6,25(2)', 'Zhongwan-6,25(3)', 'Zhongwan-6,25(4)'),
                       'Line' = c(rep('Sprint-2', 8), rep('Zhewan-1', 8), rep('Zhongwan-6', 8)),
                       'Condition' = c(rep(10, 4), rep(20, 4), rep(10, 4), rep(25, 4), rep(10, 4), rep(25, 4)))
pca_data$Color <- paste(pca_data$Line, pca_data$Condition, sep = ', ')

for (i in c(1:10)) {

iterator <- gtools::combinations(5, 2, c(1,2,3,4,5))[i,]
  
tiff(filename = file.path('revision_round1_supplementaries', 'Figure7+FigS4', paste(i, 'tiff', sep = '.')),
     width = 25, height = 20, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

g <- ggbiplot(pca_peas, choices = iterator, varname.size = 0, labels.size = 0, alpha = 0, var.axes = F, groups = pca_data$Color) +
  expand_limits(x=c(-3,1)) + theme_bw() +
  theme(text = element_text(size=13, family = 'arial', color='black'), axis.text.x = element_text(color = 'black', family = 'arial'),
        axis.text.y = element_text(color = 'black', family = 'arial'),
        axis.line.x = element_line(color = 'black'),
        legend.text = element_text(size = 13),
        legend.title =
          element_text( size = 15)) + geom_point(aes(colour = pca_data$Color), size = 4, alpha = 0.3) + xlab(paste('PC', iterator[1], sep = '')) +
  ylab(paste('PC', iterator[2], sep = ''))

print(g)

dev.off()

}

pca_eigenvectors <- data.frame('PC1' = names(abs(pca_peas$rotation[,1]) %>% sort(decreasing = T)) %>% as.character(),
                           'PC1_values' = abs(pca_peas$rotation[,1]) %>% sort(decreasing = T),
                          'PC2' = names(abs(pca_peas$rotation[,2]) %>% sort(decreasing = T)) %>% as.character(),
                           'PC2_values' = abs(pca_peas$rotation[,2]) %>% sort(decreasing = T),
                           'PC3' = names(abs(pca_peas$rotation[,3]) %>% sort(decreasing = T)) %>% as.character(),
                           'PC3_values' = abs(pca_peas$rotation[,3]) %>% sort(decreasing = T),
                           'PC4' = names(abs(pca_peas$rotation[,4]) %>% sort(decreasing = T)) %>% as.character(),
                           'PC4_values' = abs(pca_peas$rotation[,4]) %>% sort(decreasing = T),
                           'PC5' = names(abs(pca_peas$rotation[,5]) %>% sort(decreasing = T)) %>% as.character(),
                           'PC5_values' = abs(pca_peas$rotation[,5]) %>% sort(decreasing = T))

pca_eigenvectors %<>% mutate('BLASTP_names_1' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC1,][match(pca_eigenvectors$PC1, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC1,]$transcript_id),] %>% dplyr::select(BLASTP_names) %>% unlist(),
      'BLASTP_names_2' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC2,][match(pca_eigenvectors$PC2, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC2,]$transcript_id),] %>% dplyr::select(BLASTP_names) %>% unlist(),
      'BLASTP_names_3' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC3,][match(pca_eigenvectors$PC3, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC3,]$transcript_id),] %>% dplyr::select(BLASTP_names) %>% unlist(),
      'BLASTP_names_4' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC4,][match(pca_eigenvectors$PC4, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC4,]$transcript_id),] %>% dplyr::select(BLASTP_names) %>% unlist(),
      'BLASTP_names_5' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC5,][match(pca_eigenvectors$PC5, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC5,]$transcript_id),] %>% dplyr::select(BLASTP_names) %>% unlist(),
      'BLASTX_names_1' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC1,][match(pca_eigenvectors$PC1, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC1,]$transcript_id),] %>% dplyr::select(BLASTX_names) %>% unlist(),
      'BLASTX_names_2' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC2,][match(pca_eigenvectors$PC2, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC2,]$transcript_id),] %>% dplyr::select(BLASTX_names) %>% unlist(),
      'BLASTX_names_3' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC3,][match(pca_eigenvectors$PC3, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC3,]$transcript_id),] %>% dplyr::select(BLASTX_names) %>% unlist(),
      'BLASTX_names_4' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC4,][match(pca_eigenvectors$PC4, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC4,]$transcript_id),] %>% dplyr::select(BLASTX_names) %>% unlist(),
      'BLASTX_names_5' = annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC5,][match(pca_eigenvectors$PC5, annot_full_ext[annot_full_ext$transcript_id %in% pca_eigenvectors$PC5,]$transcript_id),] %>% dplyr::select(BLASTX_names) %>% unlist())

pca_eigenvectors <- pca_eigenvectors[,c(1,2,11,16,3,4,12,17,5,6,13,18,7,8,14,19,9,10,15,20)]

write.table(pca_eigenvectors, file.path('very_last_supplementary_materials', 'Table_S9.tsv'), sep = '\t', col.names = T, row.names = F)

```

### Cosine similarity metrics

```{r cosine_distances, echo=TRUE, eval=TRUE, warning=FALSE, message=FALSE}

vector_populator <- function(x, up, down) {
  if (annot_full_ext$transcript_id[x] %in% up$target_id) return(1)
  else if (annot_full_ext$transcript_id[x] %in% down$target_id) return(-1)
  else return(0)
}

vlength <- nrow(annot_full_ext)
Spri1020 <- sapply(1:vlength, function(x) vector_populator(x, up_Sprint2_com, down_Sprint2_com)) %>% `names<-`(annot_full_ext$transcript_id)
Zhe1020 <- sapply(1:vlength, function(x) vector_populator(x, up_Zhewan1, down_Zhewan1)) %>% `names<-`(annot_full_ext$transcript_id)
Zho1020 <- sapply(1:vlength, function(x) vector_populator(x, up_Zhongwan6, down_Zhongwan6)) %>% `names<-`(annot_full_ext$transcript_id)
SpriZhe10 <- sapply(1:vlength, function(x) vector_populator(x, up_10SpriZhe, down_10SpriZhe)) %>% `names<-`(annot_full_ext$transcript_id)
SpriZho10 <- sapply(1:vlength, function(x) vector_populator(x, up_10SpriZho, down_10SpriZho)) %>% `names<-`(annot_full_ext$transcript_id)
SpriZhe20 <- sapply(1:vlength, function(x) vector_populator(x, up_20SpriZhe, down_20SpriZhe)) %>% `names<-`(annot_full_ext$transcript_id)
SpriZho20 <- sapply(1:vlength, function(x) vector_populator(x, up_20SpriZho, down_20SpriZho)) %>% `names<-`(annot_full_ext$transcript_id)


cosines <- list('Spri1020/Zhe1020' = cosine(Spri1020, Zhe1020), 'Spri1020/Zho1020' = cosine(Spri1020, Zho1020),
                'Zhe1020/Zho1020' = cosine(Zhe1020, Zho1020), 'Spri1020/SpriZhe10' = cosine(Spri1020, SpriZhe10),
                'Spri1020/SpriZho10' = cosine(Spri1020, SpriZho10), 'Spri1020/SpriZhe20' = cosine(Spri1020, SpriZhe20),
                'Spri1020/SpriZho20' = cosine(Spri1020, SpriZho20))
write.table(cosines %>% `colnames<-`(sapply(colnames(cosines), function(x) gsub('\\.', '/', x))), file.path('revision_round1_supplementaries', 'cosines.tsv'), sep = '\t', col.names = T, row.names = F, dec = '.')

```

## TPM-based TF expression tracking

```{r TF heatmaps, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

plotting_matrix <- full_matrix %>% mutate('Sprint2_10_mean' = sapply(1:nrow(full_matrix), function(x) mean(as.numeric(full_matrix[x,c(2:5)]))),
                                          'Sprint2_20_mean' = sapply(1:nrow(full_matrix), function(x) mean(as.numeric(full_matrix[x,c(6:9)]))),
                                          'Zhewan1_10_mean' = sapply(1:nrow(full_matrix), function(x) mean(as.numeric(full_matrix[x,c(10:13)]))),
                                          'Zhewan1_25_mean' = sapply(1:nrow(full_matrix), function(x) mean(as.numeric(full_matrix[x,c(14:17)]))),
                                          'Zhongwan6_10_mean' = sapply(1:nrow(full_matrix), function(x) mean(as.numeric(full_matrix[x,c(18:21)]))),
                                          'Zhongwan6_25_mean' = sapply(1:nrow(full_matrix), function(x) mean(as.numeric(full_matrix[x,c(22:25)]))))

```

## Category-wise scatterplots


```{r scatter functions, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

pea_scatter_plotter <- function(plotting_matrix_loc, uprmax = NA,
                                plot_title = NA, x_breaks = c(), y_breaks = c()){

  if (!is.na(uprmax)){
    plotting_matrix_loc %<>% filter_at(vars(c(colnames(plotting_matrix_loc[,unlist(lapply(plotting_matrix_loc, is.numeric))]))), all_vars(.<=uprmax))
  }
  
  ggplot(plotting_matrix_loc, aes(x = value, y = value, size = I(2.8))) + 
    geom_point(aes(x = Sprint2_10_mean, y = Sprint2_20_mean), alpha = 0.4, col = '#068013') +
    geom_point(aes(x = Zhewan1_10_mean, y = Zhewan1_25_mean), alpha = 0.4, col = '#f1cb20') +
    geom_point(aes(x = Zhongwan6_10_mean, y = Zhongwan6_25_mean), alpha = 0.4, col = '#b30a07') +
    xlab(label = 'Earlier condition') + ylab('Latter condition')+
    guides(alpha = 'none')+
    theme_bw() + theme(plot.title = element_text(hjust = 0.5), legend.title = element_text(size = 20),
                       legend.text = element_text(size = 16), axis.title = element_text(size=14),
                       axis.text = element_text(size=12, color = 'black')) +
    scale_x_continuous(labels = scales::comma) + scale_y_continuous(labels = scales::comma) + scale_color_manual(name = 'cultivar',
          values = c('Sprint-2'='#068013', 'Zhewan-1'='#f1cb20', 'Zhongwan-6'='#b30a07'))
  
}


find_outliers <- function(df, x, y, additional_categories = NA, group, category) {
  cd <- cooks.distance(lm(get(y)~get(x), data = df))
  plot(cd,pch=1, cex = 1)
  abline(h = 4*mean(cd,na.rm = T), col = "blue")
  text(x=1:length(cd)+2, y=cd, labels=ifelse(cd>4*mean(cd, na.rm=T),names(cd),""), col="blue")
  outliers <- cd[cd>4*mean(cd, na.rm=T)] %>% names()
  df <- df[rownames(df) %in% outliers, c(x,y, additional_categories)]
  colnames(df)[c(1,2)] <- c('Early', 'Late')
  df$group <- rep(group, nrow(df)) %>% as.character()
  df$category <- rep(category, nrow(df)) %>% as.character()
  return(df)
}


```

```{r scatters}

scatter_plotnsave <- function(plotmat, filename, w=20, h=16, cutoff = NA) {
  
  tiff(filename = file.path('revision_round1_supplementaries', 'scatters', paste(filename, 'tif', sep = '.')),
     width = w, height = h, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")
  
  print(pea_scatter_plotter(plotmat, uprmax = cutoff))
  
  
  dev.off()
}

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('storage proteins', annot_full_ext$MapMan_term),]$transcript_id,], 'storage_proteins')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('storage proteins', annot_full_ext$MapMan_term),]$transcript_id,], 'storage_proteins', cutoff = 5000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(TF_GO, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('regulation of transcription', annot_full_ext$MapMan_term),]$transcript_id,], 'transcription')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(lipids, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'lipids')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(lipids, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'lipids_2k', cutoff = 2000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(translation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('protein.synthesis', annot_full_ext$MapMan_term),]$transcript_id,], 'tranaslation')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(translation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('protein.synthesis', annot_full_ext$MapMan_term),]$transcript_id,], 'tranaslation_1.6k', cutoff = 1600)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(repair, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('DNA.repair', annot_full_ext$MapMan_term),]$transcript_id,], 'repair')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(starch, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('starch', annot_full_ext$MapMan_term),]$transcript_id,], 'starch')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(starch, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('starch', annot_full_ext$MapMan_term),]$transcript_id,], 'starch_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(desiccation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'desiccation')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(postmod, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'postmod')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(degradation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('protein.degradation', annot_full_ext$MapMan_term),]$transcript_id,], 'degradation')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(degradation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('protein.degradation', annot_full_ext$MapMan_term),]$transcript_id,], 'degradation_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(replication, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'replication')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(folding, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('folding', annot_full_ext$MapMan_term),]$transcript_id,], 'folding')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(folding, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('folding', annot_full_ext$MapMan_term),]$transcript_id,], 'folding_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(chromatin, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'chromatin')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(GA, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('gibberelin', annot_full_ext$MapMan_term),]$transcript_id,], 'GA')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(ABA, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) | grepl('abscisic acid', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'ABA')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(ABA, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) | grepl('abscisic acid', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'ABA_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(splicing, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'splicing')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(cell_wall, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('cell wall', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'cell_wall')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(cell_wall, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('cell wall', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'cell_wall_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(tubulin, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'tubulin')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(motility, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'motility')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(motility, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'motility_0.2k', cutoff = 200)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(hormones, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('hormone', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'hormones')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(hormones, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('hormone', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'hormones_2k', cutoff = 2000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(PH, collapse = ')|('), ')', sep = ''),
annot_full_ext$GO_BLAST) | grepl('PS', annot_full_ext$MapMan_term),]$transcript_id,], 'PS')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(PH, collapse = ')|('), ')', sep = ''),
annot_full_ext$GO_BLAST) | grepl('PS', annot_full_ext$MapMan_term),]$transcript_id,], 'PS_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(respiration, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('(mitochondrial electron transport)|(TCA)|(glysolysis)', annot_full_ext$MapMan_term),]$transcript_id,], 'respiration')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(respiration, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('(mitochondrial electron transport)|(TCA)|(glysolysis)', annot_full_ext$MapMan_term),]$transcript_id,], 'respiration_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(redox, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('redox', annot_full_ext$MapMan_term),]$transcript_id,], 'redox')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(embryogenesis, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('embry', annot_full_ext$MapMan_term)) & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)', annot_full_ext$BLASTP_names)) & !(grepl('(legumin)|(vicilin)', annot_full_ext$BLASTX_names)),]$transcript_id,], 'enbryogenesis')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(embryogenesis, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('embry', annot_full_ext$MapMan_term)) & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)', annot_full_ext$BLASTP_names)) & !(grepl('(legumin)|(vicilin)', annot_full_ext$BLASTX_names)),]$transcript_id,], 'enbryogenesis_1k', cutoff = 1000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_develoment, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'seed_development')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_develoment, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'seed_development_3k', cutoff = 3000)

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_germination, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'seed_germination')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(meristems, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'meristems')

scatter_plotnsave(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste('(', paste(transposons, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'transposons')

```

## Tracking outliers
```{r find_outliers}

find_outliers <- function(df, x, y, additional_categories = NA, group, category) {
  cd <- cooks.distance(lm(get(y)~get(x), data = df))
  plot(cd,pch=1, cex = 1)
  abline(h = 4*mean(cd,na.rm = T), col = "blue")
  text(x=1:length(cd)+2, y=cd, labels=ifelse(cd>4*mean(cd, na.rm=T),names(cd),""), col="blue")
  outliers <- cd[cd>4*mean(cd, na.rm=T)] %>% names()
  df <- df[rownames(df) %in% outliers, c(x,y, additional_categories)]
  colnames(df)[c(1,2)] <- c('Early', 'Late')
  df$group <- rep(group, nrow(df)) %>% as.character()
  df$category <- rep(category, nrow(df)) %>% as.character()
  return(df)
}

# ABA
outlier_list <- find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(ABA, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) | grepl('abscisic acid', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'ABA metabolism and signaling')

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(ABA, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) | grepl('abscisic acid', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'ABA metabolism and signaling'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(ABA, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) | grepl('abscisic acid', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'ABA metabolism and signaling'))

# chromatin
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(chromatin, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Chromatin constituent and remodeling'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(chromatin, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Chromatin constituent and remodeling'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(chromatin, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'Chromatin constituent and remodeling'))

# degradation
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(degradation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('protein.degradation', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Protein degradation'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(degradation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('protein.degradation', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Protein degradation'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(degradation, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('protein.degradation', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'Protein degradation'))

# folding
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(folding, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)| grepl('folding', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Protein folding'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(folding, collapse = ')|('), ')', sep = ''),  annot_full_ext$GO_BLAST)| grepl('folding', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Protein folding'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(folding, collapse = ')|('), ')', sep = ''),
          annot_full_ext$GO_BLAST)| grepl('folding', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'Protein folding'))

# GA
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(GA, collapse = ')|('), ')', sep = ''),
  annot_full_ext$GO_BLAST) | grepl('gibberelin', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'GA metabolism and signaling'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(GA, collapse = ')|('), ')', sep = ''),
    annot_full_ext$GO_BLAST) | grepl('gibberelin', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'GA metabolism and signaling'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(GA, collapse = ')|('), ')', sep = ''),
            annot_full_ext$GO_BLAST) | grepl('gibberelin', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'GA metabolism and signaling'))

# lipids
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(lipids, collapse = ')|('), ')', sep = ''),
                        annot_full_ext$GO_BLAST),]$transcript_id,],  'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Lipid storage'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(lipids, collapse = ')|('), ')', sep = ''),
             annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Lipid storage'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(lipids, collapse = ')|('), ')', sep = ''),
                annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'Lipid storage'))

# post-translational modifications
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(postmod, collapse = ')|('), ')', sep = ''),
      annot_full_ext$GO_BLAST),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Post-translational modification'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(postmod, collapse = ')|('), ')', sep = ''),
            annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Post-translational modification'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(postmod, collapse = ')|('), ')', sep = ''),
            annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'Post-translational modification'))

# repair
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(repair, collapse = ')|('), ')', sep = ''),
  annot_full_ext$GO_BLAST) | grepl('DNA.repair', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'DNA repair'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(repair, collapse = ')|('), ')', sep = ''),
      annot_full_ext$GO_BLAST) | grepl('DNA.repair', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'DNA reair'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(repair, collapse = ')|('), ')', sep = ''),
      annot_full_ext$GO_BLAST) | grepl('DNA.repair', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhongwan-6', 'DNA repair'))

# replication
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(replication, collapse = ')|('), ')', sep = ''),
      annot_full_ext$GO_BLAST),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'DNA replication'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(replication, collapse = ')|('), ')', sep = ''),
 annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'DNA replication'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(replication, collapse = ')|('), ')', sep = ''),
        annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'DNA replication'))

# splicing
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(splicing, collapse = ')|('), ')', sep = ''),
    annot_full_ext$GO_BLAST),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'RNA spicing'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(splicing, collapse = ')|('), ')', sep = ''),
            annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'RNA spicing'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(splicing, collapse = ')|('), ')', sep = ''),
  annot_full_ext$GO_BLAST),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'RNA spicing'))

# starch 
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(starch, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('starch', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Starch metabolism'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(starch, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('starch', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Starch metabolism'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(starch, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('starch', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'Starch metabolism'))

# storage proteins

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),
                                                                                             annot_full_ext$GO_BLAST) | grepl('storage proteins', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Protein storage'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('storage proteins', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Protein storage'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('storage proteins', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'Protein storage'))


# transcription
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(TF_GO, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('regulation of transcription', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Transcription regulation'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(TF_GO, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('regulation of transcription', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Transcription regulation'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(TF_GO, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) | grepl('regulation of transcription', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'Transcription regulation'))

# translation
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(translation, collapse = ')|('), ')', sep = ''),
  annot_full_ext$GO_BLAST) | grepl('protein.synthesis', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Translation'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(translation, collapse = ')|('), ')', sep = ''),
        annot_full_ext$GO_BLAST) | grepl('protein.synthesis', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Translation'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(translation, collapse = ')|('), ')', sep = ''),
    annot_full_ext$GO_BLAST) | grepl('protein.synthesis', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'Translation'))


# cell wall
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(cell_wall, collapse = ')|('), ')', sep = ''),
          annot_full_ext$GO_BLAST) & !grepl('storage', annot_full_ext$MapMan_term) | grepl('cell wall', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Cell wall'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(cell_wall, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) & !grepl('storage', annot_full_ext$MapMan_term) | grepl('cell wall', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Cell wall'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(cell_wall, collapse = ')|('), ')', sep = ''),
                                                                                              annot_full_ext$GO_BLAST) & !grepl('storage', annot_full_ext$MapMan_term) | grepl('cell wall', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'Cell wall'))

# tubulin
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(tubulin, collapse = ')|('), ')', sep = ''),  annot_full_ext$GO_BLAST) & !grepl('storage', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Tubulin'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(tubulin, collapse = ')|('), ')', sep = ''),  annot_full_ext$GO_BLAST) & !grepl('storage', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Tubulin'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(tubulin, collapse = ')|('), ')', sep = ''),  annot_full_ext$GO_BLAST) & !grepl('storage', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'Tubulin'))

# tubulin-associated motility
outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(motility, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Tubulin-associated motility'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(motility, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Tubulin-associated motility'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(motility, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-6', 'Tubulin-associated motility'))

### Photosynthesis

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(PH, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('PS', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Photosynthesis'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(PH, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('PS', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Photosynthesis'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(PH, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('PS', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-1', 'Photosynthesis'))

### Respiration

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(respiration, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('(mitochondrial electron transport)|(TCA)|(glycolysis)', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Respiration'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(respiration, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('(mitochondrial electron transport)|(TCA)|(glycolysis)', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Respiration'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(respiration, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('(mitochondrial electron transport)|(TCA)|(glycolysis)', annot_full_ext$MapMan_term))  & !grepl('storage', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-1', 'Respiration'))

### Redox

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(redox, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('redox', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Redox'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(redox, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('redox', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Redox'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(redox, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('redox', annot_full_ext$MapMan_term))  & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-1', 'Redox'))

## Hormones

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(hormones, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('hormone', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Hormone metabolism and signaling'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(hormones, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('hormone', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Hormone metabolism and signaling'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[(grepl(paste0('(', paste(hormones, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) | grepl('hormone', annot_full_ext$MapMan_term)) & !grepl(paste0('(', paste(storage_proteins, collapse = ')|('), ')', sep = ''),annot_full_ext$GO_BLAST) & !grepl('storage protein', annot_full_ext$MapMan_term),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-1', 'Hormone metabolism and signaling'))

## Meristems

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(meristems, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Meristem maintenance'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(meristems, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Meristem maintenance'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(meristems, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-1', 'Meristem maintenance'))

## Seed development

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_develoment, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Seed development'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_develoment, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Seed development'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_develoment, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-1', 'Seed development'))

## Seed germination

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_germination, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Sprint2_10_mean', 'Sprint2_20_mean', c('target_id'), 'Sprint-2', 'Seed development'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_germination, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Zhewan1_10_mean', 'Zhewan1_25_mean', c('target_id'), 'Zhewan-1', 'Seed development'))

outlier_list %<>% rbind(find_outliers(plotting_matrix[plotting_matrix$target_id %in% annot_full_ext[grepl(paste0('(', paste(seed_germination, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST)  & !(grepl('storage protein', annot_full_ext$MapMan_term)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTP_names, ignore.case = T)) & !(grepl('(legumin)|(vicilin)|(lega)|(seed maturation)', annot_full_ext$BLASTX_names, ignore.case = T)),]$transcript_id,], 'Zhongwan6_10_mean', 'Zhongwan6_25_mean', c('target_id'), 'Zhewan-1', 'Seed development'))


outlier_list$'BLASTP name' <- sapply(outlier_list$target_id, function(x) annot_full_ext[annot_full_ext$transcript_id == x,]$BLASTP_names)
outlier_list %<>% mutate(Early = as.character(Early) %>% gsub(',', '.', .),
                         Late = as.character(Late) %>% gsub(',', '.', .))
outlier_list <- outlier_list[,c(5,4,3,6,1,2)]
format(outlier_list, decimal.mark = '.')
write.table(outlier_list, file.path('revision_round1_supplementaries', 'scatters',   'TableS15.tsv'), col.names = T, row.names = F,
            sep = '\t', dec = '.')

```


## Self-organizing maps

For the sake of consistency we first create a stereotypical function for SOM building:

```{r SOM_function, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

oposSOM_launch <- function(data, x,  name, wd) {
  
  previous_wd <- getwd()
  
  data <- data %>% mutate_at(vars(-target_id), all_vars(scale(.)))
  rownames(data) <- data$target_id
data %<>% dplyr::select(-c(target_id))
data[data == -Inf] <- 0
data[data == Inf] <- 0
data <- data[!(apply(data, 1, function(x) all(x == 0))),]
  
  env_som <- opossom.new(list(dataset.name = name, dim.1stLvlSom = x))
env_som$indata <- data %>% as.matrix()
env_som$group.labels <- c(rep('Sprint-2(10)', 4), rep('Sprint-2(20)', 4), rep('Zhewan-1(10)', 4), rep('Zhewan-1(25)', 4), rep('Zhongwan-6(10)', 4), rep('Zhongwan-6(25)', 4))
env_som$group.colors <- c(rep('#00b6c2', 4),rep('#0006c2', 4), rep('#dfc9e2', 4),rep('#dfe9e2', 4), rep('#228AD0', 4), rep('#228AC0', 4))
setwd(wd)
opossom.run(env = env_som)

setwd(previous_wd)

}

```



```{r oposSOMs_launch, echo=TRUE, eval=TRUE}

oposSOM_launch(full_matrix, 40, 'Full transcriptomic landscape', 'oposSOM_full')

oposSOM_launch(full_matrix %>% filter(target_id %in% annot_full_ext[grepl(paste0('(', paste(paste(TF_GO, collapse = ')|('), paste(chromatin, collapse = ')|('), paste(splicing, collapse = ')|('), 'GO:0005634', sep = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) & !grepl(paste('(', paste(storage_proteins, collapse = ')|('), ')', sep=''), annot_full_ext$GO_BLAST) ,]$transcript_id), 30, 'TFs', 'oposSOM_TFs')

oposSOM_launch(full_matrix %>% filter(target_id %in% annot_full_ext[grepl(paste0('(', paste(hormones, collapse = ')|('), ')', sep = ''), annot_full_ext$GO_BLAST) & !grepl(paste('(', paste(storage_proteins, collapse = ')|('), ')', sep=''), annot_full_ext$GO_BLAST) & !grepl(paste('(', paste('vicilin', 'legumin', sep = ')|('), ')', sep = ''), annot_full_ext$BLASTP_names, ignore.case = T),]$transcript_id), 30, 'Hormones', 'oposSOM_hormones')

oposSOM_launch(full_matrix %>% filter(target_id %in% annot_full_ext[grepl('hormone', annot_full_ext$MapMan_term) & !grepl(paste('(', paste('vicilin', 'legumin', 'lega', 'albumin', sep = ')|('), ')', sep = ''), annot_full_ext$BLASTP_names, ignore.case = T),]$transcript_id), 20, 'Hormones', 'revision_round1_supplementaries/oposSOM_hormones_MapMan_only')
cat('\014')

oposSOM_launch(full_matrix %>% filter(target_id %in% annot_full_ext[grepl('regulation of transcription', annot_full_ext$MapMan_term) & !grepl(paste('(', paste('vicilin', 'legumin', 'lega', 'albumin', sep = ')|('), ')', sep = ''), annot_full_ext$BLASTP_names, ignore.case = T),]$transcript_id), 30, 'TFs', 'revision_round1_supplementaries/oposSOM_TFs_MapMan_only')
cat('\014')

```

### Spot and Cluster Annotation

```{r SOM_spot_annotion, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

upload_spots <- function(path, prefix, pattern=NA) {
  file_list <- list.files(path=path)
  if (!is.na(pattern)) {
    file_list <- file_list[grepl(pattern, file_list)]
  }
  for (i in 1:length(file_list)) {
    name_val <- paste(prefix, 'spot', LETTERS[i], sep = '_')
    assign(name_val, read.table(file.path(path, file_list[i]), header = T, sep = ',', dec = '.'), envir = .GlobalEnv)
  }
}

TF_tracker <- function(DE_df) {
  TF_df <- annot_full_ext[grepl('regulation of transcription', annot_full_ext$MapMan_term),] %>% filter(transcript_id %in% DE_df$target_id) %>% dplyr::select(transcript_id, MapMan_term) %>% mutate(MapMan_term = MapMan_term %>% sapply(function(x) unlist(strsplit(x, ';'))[grepl('regulation of transcription', unlist(strsplit(x, ';')))])) %>% mutate(TF_Class = MapMan_term %>% sapply(function(x) x %>% strsplit(.,'\\.') %>% unlist() %>% unname() %>% nth(3)))

TF_df <- table(unlist(TF_df$TF_Class)) %>% as.data.frame() %>% arrange(desc(Freq))
  
return(TF_df)
}




```

**NOTE**: group overexpression SOM spot tables have been edited manually to avoid unnecessary tumult with table parsing in R. Name for the first column was manually set as 'target_id', and columns 'Symbol', 'Chromosome' and 'Description' were deleted from the tables.

```{r annotate_spots, echo=TRUE, eval=TRUE, message=FALSE, warning=FALSE, error=FALSE}

spot_dir <- file.path('revision_round1_supplementaries', 'SOM_spots')

upload_spots(file.path(spot_dir, 'TFs'), 'TF')
upload_spots('novel_assembly_SOM_spots/full', 'full')
upload_spots(file.path(spot_dir, 'Hormones'), 'hormone')

hormone_mapping <- create_GO_geneUniverse(annot_full_ext[grepl('hormone', annot_full_ext$MapMan_term) & !grepl(paste('(', paste('vicilin', 'legumin', 'lega', 'albumin', sep = ')|('), ')', sep = ''), annot_full_ext$BLASTP_names, ignore.case = T),], '*')
tf_mapping <- create_GO_geneUniverse(annot_full_ext[grepl('regulation of transcription', annot_full_ext$MapMan_term) & !grepl(paste('(', paste('vicilin', 'legumin', 'lega', 'albumin', sep = ')|('), ')', sep = ''), annot_full_ext$BLASTP_names, ignore.case = T),], '*')

sapply(aspects, function(x) iterating_function_GO(full_spot_A, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(full_spot_B, x, 0.01, full_mapping_ext, geneUniverse_ext))
sapply(aspects, function(x) iterating_function_GO(full_spot_C, x, 0.01, full_mapping_ext, geneUniverse_ext))

sapply(aspects, function(x) iterating_function_GO(TF_spot_A, x, 0.01, tf_mapping, names(tf_mapping)))
sapply(aspects, function(x) iterating_function_GO(TF_spot_B, x, 0.01, tf_mapping, names(tf_mapping)))
sapply(aspects, function(x) iterating_function_GO(TF_spot_C, x, 0.01, tf_mapping, names(tf_mapping)))
sapply(aspects, function(x) iterating_function_GO(TF_spot_D, x, 0.01, tf_mapping, names(tf_mapping)))

sapply(aspects, function(x) iterating_function_GO(hormone_spot_A, x, 0.01, hormone_mapping, names(hormone_mapping)))
sapply(aspects, function(x) iterating_function_GO(hormone_spot_B, x, 0.01, hormone_mapping, names(hormone_mapping)))
sapply(aspects, function(x) iterating_function_GO(hormone_spot_C, x, 0.01, hormone_mapping, names(hormone_mapping)))
sapply(aspects, function(x) iterating_function_GO(hormone_spot_D, x, 0.01, hormone_mapping, names(hormone_mapping)))



sapply(ls()[grepl('((BP)|(CC)|(MF))_hormone_spot', ls())], function(x) x %>% get() %>% 
         write.table(file.path('revision_round1_supplementaries', 'SOM_spots', 'Hormones', paste(x, 'tsv', sep = '.')),
                     sep = '\t', row.names = F, col.names = T))

sapply(ls()[grepl('^(((BP)|(CC)|(MF))_)?TF_spot', ls())], function(x) x %>% get() %>% 
         write.table(file.path('revision_round1_supplementaries', 'SOM_spots', 'TFs', paste(x, 'tsv', sep = '.')),
                     sep = '\t', row.names = F, col.names = T))

spot_dir <- 'hormone_spots_291219'
upload_spots(spot_dir, 'hormones_mapman')
hormones_mapman_mapping <- create_GO_geneUniverse(annot_full_ext %>% filter(grepl('hormone', MapMan_term)), '')

sapply(aspects, function(x) iterating_function_GO(hormones_mapman_spot_A, x, 0.01, hormones_mapman_mapping, names(hormones_mapman_mapping)))
hormones_mapman_spot_B %<>% dplyr::rename(target_id = target_Id)
sapply(aspects, function(x) iterating_function_GO(hormones_mapman_spot_B, x, 0.01, hormones_mapman_mapping, names(hormones_mapman_mapping)))
sapply(aspects, function(x) iterating_function_GO(hormones_mapman_spot_C, x, 0.01, hormones_mapman_mapping, names(hormones_mapman_mapping)))
sapply(aspects, function(x) iterating_function_GO(hormones_mapman_spot_D, x, 0.01, hormones_mapman_mapping, names(hormones_mapman_mapping)))

TF_SOM_MM <- merge(TF_tracker(TF_spot_A), TF_tracker(TF_spot_B), by = 'Var1', all.x = T, all.y = T) %>% merge(TF_tracker(TF_spot_C), by = 'Var1', all.x = T, all.y = T) %>% merge(TF_tracker(TF_spot_D), by = 'Var1', all.x = T, all.y = T) %>% set_colnames(c('TF_family', 'A', 'B', 'C', 'D')) %>% mutate_at(vars(-c(1)), ~replace(., is.na(.), 0))%>% arrange(desc(A)) %>% column_to_rownames('TF_family')

tiff(filename = file.path('revision_round1_supplementaries', 'Figures', 'Fig11', '2.tiff'),
     width = 20, height = 30, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

pheatmap(TF_SOM_MM %>% as.matrix(), color = colors[c(seq(1, 200, by = 5), seq(201, 255))], family = 'arial')

dev.off()

```

## SNP Analysis

```{r}

snp_analysis_processing <- function(path) {
  frame <- read.table(path, header = F, sep = '\t', stringsAsFactors = F) %>% dplyr::rename(target_id = V1) %>% filter(!grepl('LOW', V8))
  frame$Moderate <- ifelse(grepl('MODERATE', frame$V8), 1, 0)
  frame$High <- ifelse(grepl('HIGH', frame$V8), 1, 0)
  frame$Modifier <- ifelse(grepl('MODIFIER', frame$V8), 1, 0)
  frame$identifier <- sapply(1:nrow(frame), function(x) frame[x,c(1,2,4,5)] %>% as.vector() %>% paste(collapse = ''))
return(frame)
  }

Sprint2_snp_het <- snp_analysis_processing(file.path('for_further_analysis', 'Sprint2_snps_annotated_no_err_Het.vcf'))
Zhewan1_snp_het <- snp_analysis_processing(file.path('for_further_analysis', 'Zhewan1_snps_annotated_no_err_Het.vcf'))
Zhewan1_snp_hom <- snp_analysis_processing(file.path('for_further_analysis', 'Zhewan1_snps_annotated_no_err_Homo.vcf'))
Zhongwan6_snp_het <- snp_analysis_processing(file.path('for_further_analysis', 'Zhongwan6_snps_annotated_no_err_Het.vcf'))
Zhongwan6_snp_hom <- snp_analysis_processing(file.path('for_further_analysis', 'Zhongwan6_snps_annotated_no_err_Homo.vcf'))

all_hets <- rbind(Sprint2_snp_het, Zhewan1_snp_het, Zhongwan6_snp_het)
all_hets <- all_hets[!duplicated(all_hets$identifier),] %>% filter(!grepl('LOW', V8))
all_hets$'Zhewan-1' <- ifelse(all_hets$identifier %in% Zhewan1_snp_het$identifier, 1, 0)
all_hets$'Zhongwan-6' <- ifelse(all_hets$identifier %in% Zhongwan6_snp_het$identifier, 1, 0)
all_hets$'Sprint-2' <- ifelse(all_hets$identifier %in% Sprint2_snp_het$identifier, 1, 0)
all_hets <- all_hets[,c(1,11,12,13, 15, 16, 17)]
colnames(all_hets)[1] <- 'Transcript'

tiff(filename = file.path('revision_round1_supplementaries', 'SNP_upsets', 'all_hets.tiff'),
     width = 40, height = 20, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")

upset(all_hets, sets = c('Zhewan-1', 'Zhongwan-6', 'Sprint-2', 'Moderate', 'High', 'Modifier'), sets.bar.color = '#6dc585',
      order.by = 'freq', text.scale = c(2.1, 2, 1.8, 1.5, 2, 2), mainbar.y.label = 'Number of SNPs')

dev.off()

all_homs <- rbind(Sprint2_snp_het, Zhewan1_snp_hom, Zhongwan6_snp_hom)
all_homs <- all_homs[!duplicated(all_homs$identifier),] %>% filter(!grepl('LOW', V8))
all_homs$'Zhewan-1' <- ifelse(all_homs$identifier %in% Zhewan1_snp_hom$identifier, 1, 0)
all_homs$'Zhongwan-6' <- ifelse(all_homs$identifier %in% Zhongwan6_snp_hom$identifier, 1, 0)
all_homs$'Sprint-2' <- ifelse(all_homs$identifier %in% Sprint2_snp_het$identifier, 1, 0)
all_homs <- all_homs[,c(1,11,12,13, 15, 16, 17)]
colnames(all_homs)[1] <- 'Transcript'

tiff(filename = file.path('revision_round1_supplementaries', 'SNP_upsets', 'all_homs.tiff'),
     width = 40, height = 20, units = "cm", pointsize = 12,
     compression = "lzw",
     bg = "white", res = 520, family = "",
     type = "cairo")
 

upset(all_homs, sets = c('Zhewan-1', 'Zhongwan-6', 'Sprint-2', 'Moderate', 'High', 'Modifier'), sets.bar.color = '#6dc585',
      order.by = 'freq', text.scale = c(2.1, 2, 1.8, 1.5, 2, 2), mainbar.y.label = 'Number of SNPs')

dev.off()


Sprint2_snp_het_dedup <- Sprint2_snp_het %>% filter(!duplicated(target_id))
sapply(aspects, function(x) iterating_function_GO(Sprint2_snp_het_dedup, x, cutoff = 0.01))
sapply(aspects, function(x) iterating_function_GO(Sprint2_snp_het, x, cutoff = 0.01))

Zhewan1_snp_het_dedup <- Zhewan1_snp_het %>% filter(!duplicated(target_id))
sapply(aspects, function(x) iterating_function_GO(Zhewan1_snp_het_dedup, x, cutoff = 0.01))
sapply(aspects, function(x) iterating_function_GO(Zhewan1_snp_het, x, cutoff = 0.01))

Zhewan1_snp_hom_dedup <- Zhewan1_snp_hom %>% filter(!duplicated(target_id))
sapply(aspects, function(x) iterating_function_GO(Zhewan1_snp_hom_dedup, x, cutoff = 0.01))
sapply(aspects, function(x) iterating_function_GO(Zhewan1_snp_hom, x, cutoff = 0.01))
BP_Zhewan1_snp_hom %>% View()

Zhongwan6_snp_het_dedup <- Zhongwan6_snp_het %>% filter(!duplicated(target_id))
sapply(aspects, function(x) iterating_function_GO(Zhongwan6_snp_het_dedup, x, cutoff = 0.01))
sapply(aspects, function(x) iterating_function_GO(Zhongwan6_snp_het, x, cutoff = 0.01))

Zhongwan6_snp_hom_dedup <- Zhongwan6_snp_hom %>% filter(!duplicated(target_id))
sapply(aspects, function(x) iterating_function_GO(Zhongwan6_snp_hom_dedup, x, cutoff = 0.01))
sapply(aspects, function(x) iterating_function_GO(Zhongwan6_snp_hom, x, cutoff = 0.01))


sapply(ls()[grepl('((BP)|(CC)|(MF)).*_dedup', ls())], function(x) x %>% get() %>%  
       write.table(., file.path('VC_GO_041219', 'transcript_wise', paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))

sapply(ls()[grepl('((BP)|(CC)|(MF))_Zhewan1_snp_hom_dedup', ls())], function(x) x %>% get() %>%  
       write.table(., file.path('VC_GO_041219', 'transcript_wise', paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))

sapply(ls()[grepl('((BP)|(CC)|(MF)){1}.*_snp_((hom)|(het))$', ls(), perl = T)], function(x) x %>% get() %>% 
         write.table(.,file.path('VC_GO_041219', 'variant_wise',  paste(x, 'tsv', sep = '.')), sep = '\t', col.names = T, row.names = F))


snp_barplots <- function(df, cutoff = 10) {
  
  combined_dataset <- df[cutoff,] %>% arrange(desc(Percentage_of_Differentially_Expressed_Genes_from_Number_Of_Annotated_Genes)) %>% mutate(Term = factor(Term, levels = unique(Term))) %>% filter_all(all_vars(!is.na(.)))
  ggplot(df, aes(y=Percentage_of_Differentially_Expressed_Genes_from_Number_Of_Annotated_Genes,
                                                               x=reorder(Term, Percentage_of_Differentially_Expressed_Genes_from_Number_Of_Annotated_Genes, max))) + 
    geom_bar(stat = 'identity', position = 'identity', width = 0.8, fill = '#068013') + coord_flip() +
    theme(text = element_text(size=24, family = 'arial', color='black'), axis.text.x = element_text(color = 'black', family = 'arial'),
          axis.title.x = element_text(color = 'black', family = 'arial', size = 20),
          axis.text.y = element_text(color = 'black', family = 'arial', hjust = 1),
          axis.title.y = element_blank(),
          axis.line.x = element_line(color = 'black'),
          axis.ticks.length.y =  unit(0, 'lines'),
          panel.background = element_blank(), legend.text = element_text(size = 22), legend.title = 
            element_text(size = 28)) + 
    labs(y = '% of annotated genes')
  
}

bars_plot_n_save_snp <- function(data, cutoff, path, w=48, h=20) {
 tiff(filename = path,
      width = w, height = h, units = "cm", pointsize = 12,
      compression = "lzw",
      bg = "white", res = 520, family = "",
      type = "cairo") 

  print(snp_barplots(data, cutoff))
  
  dev.off()
    
}

snp_barplots(BP_Sprint2_snp_het, 10)

bars_plot_n_save_snp(BP_Sprint2_snp_het, cutoff = 20, path=file.path('revision_round1_supplementaries', 'snp_GO', 'A.tiff'), w = 80)
bars_plot_n_save_snp(CC_Sprint2_snp_het, cutoff = 20, path=file.path('revision_round1_supplementaries', 'snp_GO', 'B.tiff'))
bars_plot_n_save_snp(MF_Sprint2_snp_het, cutoff = 20, path=file.path('revision_round1_supplementaries', 'snp_GO', 'C.tiff'))


```
### SNP Categorical Subsetting

```{r high_moderate_modifier, echo=T, eval=T, message=F, warning=F, error=F}

snp_matcher <- function(dataset) {
  annotations <- annot_full_ext %>% filter(transcript_id %in% dataset$target_id) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names)
  return(cbind(dataset[,1:9], annotations[match(dataset$target_id, annotations$transcript_id), -1]))
}

HIGH_Sprint2 <- Sprint2_snp_het %>% filter(High == 1)
sapply(aspects, function(x) iterating_function_GO(HIGH_Sprint2, x, 0.01, full_mapping, geneUniverse))
HIGH_Sprint2 <- snp_matcher(HIGH_Sprint2)

MODERATE_Sprint2 <- snp_matcher(Sprint2_snp_het %>% filter(Moderate == 1))
sapply(aspects, function(x) iterating_function_GO(MODERATE_Sprint2, x, 0.01, full_mapping, geneUniverse))

MODIFIER_Sprint2 <- snp_matcher(Sprint2_snp_het %>% filter(Modifier == 1))
sapply(aspects, function(x) iterating_function_GO(MODIFIER_Sprint2, x, 0.01, full_mapping, geneUniverse))

HIGH_Zhewan1_hom <- snp_matcher(Zhewan1_snp_hom %>% filter(High == 1))
sapply(aspects, function(x) iterating_function_GO(HIGH_Zhewan1_hom, x, 0.01, full_mapping, geneUniverse))

MODERATE_Zhewan1_hom <- snp_matcher(Zhewan1_snp_hom %>% filter(Moderate == 1))
sapply(aspects, function(x) iterating_function_GO(MODERATE_Zhewan1_hom, x, 0.01, full_mapping, geneUniverse))

MODIFIER_Zhewan1_hom <- snp_matcher(Zhewan1_snp_hom %>% filter(Modifier == 1))
sapply(aspects, function(x) iterating_function_GO(MODIFIER_Zhewan1_hom, x, 0.01, full_mapping, geneUniverse))


HIGH_Zhongwan6_hom <- snp_matcher(Zhongwan6_snp_hom %>% filter(High == 1))
sapply(aspects, function(x) iterating_function_GO(HIGH_Zhongwan6_hom, x, 0.01, full_mapping, geneUniverse))

MODERATE_Zhongwan6_hom <- snp_matcher(Zhongwan6_snp_hom %>% filter(Moderate == 1))
sapply(aspects, function(x) iterating_function_GO(MODERATE_Zhongwan6_hom, x, 0.01, full_mapping, geneUniverse))

MODIFIER_Zhongwan6_hom <- snp_matcher(Zhongwan6_snp_hom %>% filter(Modifier == 1))
sapply(aspects, function(x) iterating_function_GO(MODIFIER_Zhongwan6_hom, x, 0.01, full_mapping, geneUniverse))

HIGH_Zhewan1_het <- snp_matcher(Zhewan1_snp_het %>% filter(High == 1))
sapply(aspects, function(x) iterating_function_GO(HIGH_Zhewan1_het, x, 0.01, full_mapping, geneUniverse))

MODERATE_Zhewan1_het <- snp_matcher(Zhewan1_snp_het %>% filter(Moderate == 1))
sapply(aspects, function(x) iterating_function_GO(MODERATE_Zhewan1_het, x, 0.01, full_mapping, geneUniverse))

MODIFIER_Zhewan1_het <- snp_matcher(Zhewan1_snp_het %>% filter(Modifier == 1))
sapply(aspects, function(x) iterating_function_GO(MODIFIER_Zhewan1_het, x, 0.01, full_mapping, geneUniverse))

HIGH_Zhongwan6_het <- snp_matcher(Zhongwan6_snp_het %>% filter(High == 1))
sapply(aspects, function(x) iterating_function_GO(HIGH_Zhongwan6_het, x, 0.01, full_mapping, geneUniverse))

MODERATE_Zhongwan6_het <- snp_matcher(Zhongwan6_snp_het %>% filter(Moderate == 1))
sapply(aspects, function(x) iterating_function_GO(MODERATE_Zhongwan6_het, x, 0.01, full_mapping, geneUniverse))

MODIFIER_Zhongwan6_het <- snp_matcher(Zhongwan6_snp_het %>% filter(Modifier == 1))
sapply(aspects, function(x) iterating_function_GO(MODIFIER_Zhongwan6_het, x, 0.01, full_mapping, geneUniverse))


sapply(ls()[grep('(HIGH)|(MODIFIER)|(MODERATE)', ls())], function(x) x %>% get() %>% 
         write.table(file.path('very_last_supplementary_materials', 'snp_supplementaries', paste(x, 'tsv', sep = '.')),
                     sep = '\t', col.names = T, row.names = F))

snp_matcher(Sprint2_snp_het %>% filter(Moderate == 1, identifier %in% Zhewan1_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Sprint_Zhewan_moderate_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Sprint2_snp_het %>% filter(Moderate == 1, identifier %in% Zhongwan6_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Sprint_Zhongwan_moderate_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Zhewan1_snp_hom %>% filter(Moderate == 1, identifier %in% Zhongwan6_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Zhewan_Zhongwan_moderate_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Sprint2_snp_het %>% filter(Modifier == 1, identifier %in% Zhewan1_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Sprint_Zhewan_modifier_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Sprint2_snp_het %>% filter(Modifier == 1, identifier %in% Zhongwan6_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Sprint_Zhongwan_modifier_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Zhewan1_snp_hom %>% filter(Modifier == 1, identifier %in% Zhongwan6_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Zhewan_Zhongwan_modifier_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Sprint2_snp_het %>% filter(High == 1, identifier %in% Zhewan1_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Sprint_Zhewan_high_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Sprint2_snp_het %>% filter(High == 1, identifier %in% Zhongwan6_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Sprint_Zhongwan_high_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

snp_matcher(Zhewan1_snp_hom %>% filter(High == 1, identifier %in% Zhongwan6_snp_hom$identifier)) %>% write.table(
  file.path('very_last_supplementary_materials', 'snp_supplementaries', 'common_snp', 'Zhewan_Zhongwan_high_common.tsv'), col.names = T, 
  row.names = F, sep = '\t')

sapply(ls()[grepl('((BP)|(CC)|(MF))_((HIGH)|(MODERATE)|(MODIFIER)).*_het', ls())], function(x) get(x) %>% 
         write.table(file.path('very_last_supplementary_materials', 'snp_supplementaries', paste(x, 'tsv', sep = '.')),
                     sep = '\t', col.names = T, row.names = F))

common_high <- snp_matcher(Zhewan1_snp_hom %>% filter(High == 1, identifier %in% Zhongwan6_snp_hom$identifier))
common_moderate <- snp_matcher(Zhewan1_snp_hom %>% filter(Moderate == 1, identifier %in% Zhongwan6_snp_hom$identifier))
common_modifier <- snp_matcher(Zhewan1_snp_hom %>% filter(Modifier == 1, identifier %in% Zhongwan6_snp_hom$identifier))

all_homs %>% filter(`Zhewan-1` == 1 & `Zhongwan-6` == 1, Moderate == 1) %>% nrow()

```

## Particular transcript tracking

```{r particular_transcripts}

ABI5 <- annot_full_ext %>% filter(grepl('abscisic acid-insensitive 5', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

SPDSY <- annot_full_ext %>% filter(grepl('spermi.*ne synthase', BLASTP_names, ignore.case = T, perl = T)) %>% dplyr::select(transcript_id) %>%  unlist() %>% unname()

FUS3 <- annot_full_ext %>% filter(grepl('fus3', BLASTX_names, ignore.case = T, perl = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

ABI3 <- annot_full_ext %>% filter(grepl('abi3', BLASTP_names, ignore.case = T, perl = T)) %>% dplyr::select(transcript_id) %>%  unlist() %>% unname() %>% dplyr::nth(1)

LEC1 <- 'TRINITY_DN7488_c0_g2_i1' # BLAST-estimated

LEC2 <- annot_full_ext %>% filter(grepl('LEC2', BLASTP_names, ignore.case = T, perl = T)) %>% dplyr::select(transcript_id) %>%  unlist() %>% unname()

DICER <- annot_full_ext %>% filter(grepl('dicer', BLASTP_names, ignore.case = T, perl = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

DOG1 <- annot_full_ext %>% filter(grepl('dog1', BLASTP_names, ignore.case = T, perl = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

GA3ox <- annot_full_ext %>% filter(grepl('gibberellin 3', BLASTP_names, ignore.case = T, perl = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()
  
DRM <- annot_full_ext %>% filter(grepl('drm', BLASTP_names, ignore.case = T), grepl('cytosine', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

CMT <- annot_full_ext %>% filter(grepl('cmt', BLASTP_names, ignore.case = T), !grepl('rubisco', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

Pol45 <- annot_full_ext %>% filter(grepl('GO:0000418', GO_BLAST)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

PCNA <- annot_full_ext %>% filter(grepl('PROLIFERATING CELL NUCLEAR ANTIGEN', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

SuSy <- annot_full_ext %>% filter(grepl('sucrose synthase', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

starch_synthase <- annot_full_ext %>% filter(grepl('starch synthase', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

AGPase <- annot_full_ext %>% filter(grepl('ADP-glucose pyrophosphorylase', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

SBE <- annot_full_ext %>% filter(grepl('starch branching enzyme', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

SDE <- annot_full_ext %>% filter(grepl('starch debranching enzyme', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

DDM1 <- annot_full_ext %>% filter(grepl('ddm1', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

RDM1 <- 'TRINITY_DN6525_c0_g1_i5'

all_transferases <- annot_full_ext %>% filter(grepl('dna.*methyltransferase', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

SAdeMetSynt <- annot_full_ext %>% filter(grepl('S-adenosylmethionine synthase', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

MetSynt <- annot_full_ext %>% filter(grepl(' methionine synthase', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

LEA <- annot_full_ext %>% filter(grepl('late embryogenesis', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()

IDN2 <- annot_full_ext %>% filter(grepl('involved in de novo', BLASTP_names, ignore.case = T)) %>% dplyr::select(transcript_id) %>% unlist() %>% unname()



# Uniparental expression of Pol IV-dependent siRNAs in developing endosperm of Arabidopsis

up_10SpriZhe %>% filter(target_id %in% DDM1)


transposons_up_Sprint2 <- annot_full %>% filter(grepl(paste('(', paste(transposons, collapse =  ')|('), ')'), GO_BLAST), transcript_id %in% up_Sprint2$target_id) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names)

transposons_down_Sprint2 <- annot_full %>% filter(grepl(paste('(', paste(transposons, collapse =  ')|('), ')'), GO_BLAST), transcript_id %in% down_Sprint2$target_id) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names)

transposons_10Zhe <- annot_full_ext %>% filter(grepl(paste('(', paste(transposons, collapse =  ')|('), ')'), GO_BLAST), transcript_id %in% down_10SpriZhe$target_id) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names)

transposons_10Zho <- annot_full_ext %>% filter(grepl(paste('(', paste(transposons, collapse =  ')|('), ')'), GO_BLAST), transcript_id %in% down_10SpriZho$target_id) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names)

transposons_20Zhe <- annot_full_ext %>% filter(grepl(paste('(', paste(transposons, collapse =  ')|('), ')'), GO_BLAST), transcript_id %in% down_20SpriZhe$target_id) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names)

transposons_20Zho <- annot_full_ext %>% filter(grepl(paste('(', paste(transposons, collapse =  ')|('), ')'), GO_BLAST), transcript_id %in% down_20SpriZho$target_id) %>% dplyr::select(transcript_id, BLASTP_names, BLASTX_names)

```


